
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Data Management and Visualization &#8212; VectorBiTE 2021</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="How to Digitize Data from Graphs in Published Papers" href="2-Digitizer_explanation.html" />
    <link rel="prev" title="Welcome to the VectorBiTE 2021 Training Workshop!" href="0-intro.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">VectorBiTE 2021</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="0-intro.html">
   Welcome to the VectorBiTE 2021 Training Workshop!
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Data Wrangling
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Data Management and Visualization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="2-Digitizer_explanation.html">
   How to Digitize Data from Graphs in Published Papers
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Model Fitting
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="3-ModelFitting-NLLS.html">
   Model Fitting using Non-linear Least-squares
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="4-ModelFitting-MLE.html">
   Model Fitting using Maximum Likelihood
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="5-ModelFitting-Bayesian.html">
   Model Fitting the Bayesian Way
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="6-Bootstrap_rTPC_tutorial_Nesbit.html">
   Bootstrapping using rTPC package
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/notebooks/1-Data_R.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/nesbitm/Vector_BiTE_21"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/nesbitm/Vector_BiTE_21/issues/new?title=Issue%20on%20page%20%2Fnotebooks/1-Data_R.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/nesbitm/Vector_BiTE_21/master?urlpath=tree/notebooks/1-Data_R.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#r-vs-python">
     R vs. Python
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-wrangling">
   Data wrangling
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#some-data-wrangling-principles">
   Some data wrangling principles
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#an-example">
     An example
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#keep-a-metadata-file-for-each-unique-dataset">
     Keep a metadata file for each unique dataset
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#minimize-modifying-raw-data-by-hand">
     Minimize modifying raw data by hand
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#convert-wide-format-data-to-long-format-using-scripts">
     Convert wide format data to long format using scripts
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#on-to-data-exploration">
   On to data exploration
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#and-then-came-tidyverse">
     And then came
     <code class="docutils literal notranslate">
      <span class="pre">
       tidyverse
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#practicals">
   Practicals
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-wrangling-with-tidyverse">
     Data wrangling with
     <code class="docutils literal notranslate">
      <span class="pre">
       tidyverse
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#handling-big-data-in-r">
   Handling Big Data in R
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#databases-and-r">
     Databases and R
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-visualization">
   Data visualization
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-exploration-with-basic-plotting">
     Data exploration with basic plotting
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#basic-plotting-commands-in-r">
       Basic plotting commands in R
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#scatter-plots">
     Scatter Plots
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#histograms">
     Histograms
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#exercise">
       Exercise
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#subplots">
     Subplots
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#overlaying-plots">
     Overlaying plots
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#boxplots">
     Boxplots
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#combining-plot-types">
     Combining plot types
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#saving-your-graphics">
     Saving your graphics
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id1">
   Practicals
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#body-mass-distributions">
     Body mass distributions
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#beautiful-graphics-in-r">
   Beautiful graphics in R
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#quick-plotting-with-qplot">
     Quick plotting with
     <code class="docutils literal notranslate">
      <span class="pre">
       qplot
      </span>
     </code>
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#scatterplots">
       Scatterplots
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#aesthetic-mappings">
       Aesthetic mappings
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#setting-transparency">
       Setting transparency
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#adding-smoothers-and-regression-lines">
       Adding smoothers and regression lines
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#id2">
       Boxplots
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#histograms-and-density-plots">
       Histograms and density plots
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#multi-faceted-plots">
       Multi-faceted plots
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#logarithmic-axes">
       Logarithmic axes
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#plot-annotations">
       Plot annotations
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#saving-your-plots">
       Saving your plots
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#some-more-important-ggplot-options">
     Some more important ggplot options
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-geom-argument">
     The
     <code class="docutils literal notranslate">
      <span class="pre">
       geom
      </span>
     </code>
     argument
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#advanced-plotting-ggplot">
     Advanced plotting:
     <code class="docutils literal notranslate">
      <span class="pre">
       ggplot
      </span>
     </code>
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#some-useful-ggplot-examples">
     Some useful ggplot examples
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plotting-a-matrix">
     Plotting a matrix
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plotting-two-dataframes-together">
     Plotting two dataframes together
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#id3">
       Exercise
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#annotating-plots">
     Annotating plots
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id4">
   Practicals
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mathematical-display">
   Mathematical display
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ggthemes">
   ggthemes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id5">
   Practicals
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plotting-script">
     Plotting script
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#visualizing-regression-analyses">
     Visualizing Regression analyses
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mapping">
     Mapping
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#groupwork-practical-regression-analysis">
     Groupwork practical: Regression analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#readings-resources">
   Readings &amp; Resources
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-management-reformatting-and-cleaning">
     Data management, reformatting and cleaning
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#visualization">
     Visualization
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="data-management-and-visualization">
<h1>Data Management and Visualization<a class="headerlink" href="#data-management-and-visualization" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><p><em>Clutter and confusion are failures of design, not attributes of information.</em> – Edward Tufte</p>
</div></blockquote>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This chapter aims at introducing you to key principles and methods for data processing, storage, exploration and visualization.</p>
<p>In this modern world, massive amounts of data are being generated in biology due to rapid advances in technologies for collecting new, as well as for digitizing old data. Some prominent examples are <em>Genomic</em>, <em>Ecosystem respiration</em>, <em>Climatic</em>, and <em>Animal tracking</em> data. Ultimately, the goal of quantitative biology is to both, discover patterns in these data, and fit mathematical models to them. Reproducible data manipulation, analyses and visualization are particularly necessary when data are so large and complex, and few are more so than biological data, which are extremely heterogeneous in their structure and quality.  Furthermore, when these data are “big” (more below on what makes a dataset “big”), computationally-efficient data handling, manipulation and analysis techniques are needed.</p>
<div class="section" id="r-vs-python">
<h3>R vs. Python<a class="headerlink" href="#r-vs-python" title="Permalink to this headline">¶</a></h3>
<p>We will use R in this chapter because it a great one stop solution for both data manipulation, analysis and visualization. If you are keen on trying data analyses in Python, see <span class="xref myst">this Appendix</span>. In general, R will do the job for most of your purposes. There not much between these two languages for data science. One advantage that Python has is its greater computational efficiency&amp;mdash. On the other hand, R was developed for convenient statistical analyses, with efficiency not being the main concern.</p>
<p>Read more about R vs Python for data science <a class="reference external" href="https://www.kdnuggets.com/2015/05/r-vs-python-data-science.html">here</a> and <a class="reference external" href="http://res.cloudinary.com/dyd911kmh/image/upload/f_auto,q_auto:best/v1523009719/main-qimg-9dcf536c501455f073dfbc4e09798a51_vpijr0.png">here</a>.</p>
</div>
</div>
<div class="section" id="data-wrangling">
<h2>Data wrangling<a class="headerlink" href="#data-wrangling" title="Permalink to this headline">¶</a></h2>
<p>You are likely to spend far more time than you think dredging through data files manually – checking them, editing them, and reformatting them to make them useful for data exploration and analysis. It is often the case that you’ll have to deal with messy or incomplete data, either because of sampling challenges (e.g., “field” data), or because you got given data that was poorly recorded and maintained. The data we obtain from different data sources is often unusable at the beginning; for example you may need to:</p>
<ul class="simple">
<li><p>Identify the variables vs observations within the data—somebody else might have recorded the data, or you youself might have collected the data some time back!</p></li>
<li><p>Fill in zeros (true measured or observed absences)</p></li>
<li><p>Identify and add a value (e.g., <code class="docutils literal notranslate"><span class="pre">-999999</span></code>) to denote missing observations</p></li>
<li><p>Derive or calculate new variables from the raw observations (e.g., convert measurements to SI units; kilograms, meters, seconds, etc.)</p></li>
<li><p>Reshape/reformat your data into a layout that works best for analysis (e.g., for R itself);e.g., from wide to long data format for replicated (across plates, chambers, plots, sites, etc.) data</p></li>
<li><p>Merge multiple datasets together into a single data sheet</p></li>
</ul>
<p>This is not an exhaustive list. Doing so many different things to your raw data is both time-consuming and risky. Why risky? Because  to err is very human, and every new, tired mouse-click and/or keyboard-stab has a high probability of inducing an erroneous data point!</p>
<p><img alt="image" src="../_images/Wrangling2.jpg" />
<small> <center>(A (metaphorical) datum being wrangled into submission <br> Source: <a class="reference external" href="https://pathanruet.files.wordpress.com/2012/05/unix-tree.png">https://pathanruet.files.wordpress.com/2012/05/unix-tree.png</a>)</center></small></p>
</div>
<div class="section" id="some-data-wrangling-principles">
<h2>Some data wrangling principles<a class="headerlink" href="#some-data-wrangling-principles" title="Permalink to this headline">¶</a></h2>
<p>So you would like a record of the data wrangling process (so that it  is repeatable and even reversible), and automate it to the extent  possible. To this end, here are some guidelines:</p>
<ul class="simple">
<li><p>Store data in universally (machine)-readable, non-proprietary formats; basically, use plain ASCII text for your file names, variable/field/column names, and data values. And make sure the data file’s “text encoding” is correct and standard  (e.g., <code class="docutils literal notranslate"><span class="pre">UTF-8</span></code>).</p></li>
<li><p>Keep a metadata file for each unique dataset (again, in  non-proprietary format).</p></li>
<li><p>Minimize modifying raw data by hand—use scripts  instead—keep a copy of the data as they were recorded.</p></li>
<li><p>Use meaningful names for your data and files and field (column) names</p></li>
<li><p>When you add data, try not to add columns (widening the format); rather, design your tables/data-sheets so that you add only rows (lengthening the format)—and convert “wide format data” to “long format data” using  scripts, not by hand,</p></li>
<li><p>All cells within a data column should contain only one type of information (i.e., either text (character), numeric, etc.).</p></li>
<li><p>Ultimately, consider creating a relational database for your data (More on this below).</p></li>
</ul>
<p>This is not an exhaustive list either— see the <a class="reference external" href="#Readings-&amp;-Resources">Readings &amp; Resources Section</a>.</p>
<div class="section" id="an-example">
<h3>An example<a class="headerlink" href="#an-example" title="Permalink to this headline">¶</a></h3>
<p>We will use the Pound Hill dataset collected by students in a past Silwood Field Course for understanding some of these  principles.</p>
<p>To start with, we need to import the <em>raw</em> data file, for which, follow these steps:</p>
<p>★ Copy the file <code class="docutils literal notranslate"><span class="pre">PoundHillData.csv</span></code> and <code class="docutils literal notranslate"><span class="pre">PoundHillMetaData.csv</span></code> files from TheMulQuaBio’s <code class="docutils literal notranslate"><span class="pre">data</span></code> directory into your own R week’s <code class="docutils literal notranslate"><span class="pre">data</span></code> directory. Then load the data in R:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyData</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;../data/PoundHillData.csv&quot;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">))</span>
<span class="nf">class</span><span class="p">(</span><span class="n">MyData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'matrix'</li><li>'array'</li></ol>
</div></div>
</div>
<ul class="simple">
<li><p>Loading the data <code class="docutils literal notranslate"><span class="pre">as.matrix()</span></code>, and setting <code class="docutils literal notranslate"><span class="pre">header=FALSE</span></code> guarantees that the data are imported “as is” so that you can wrangle them. Otherwise <code class="docutils literal notranslate"><span class="pre">read.csv</span></code> will convert the first row to column headers.</p></li>
<li><p>All the data will be converted to the character class in the resulting matrix called <code class="docutils literal notranslate"><span class="pre">MyData</span></code> because at least one of the entries is already character class.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As of R version 4.0.0 of R released in April 2020, the default for
<code class="docutils literal notranslate"><span class="pre">stringsAsFactors</span></code> <a class="reference external" href="https://cran.r-project.org/doc/manuals/r-devel/NEWS.html">was changed to false</a> If you are using R version 3.x.x, you will need to add <code class="docutils literal notranslate"><span class="pre">stringsAsFactors</span> <span class="pre">=</span> <span class="pre">FALSE</span></code> to the above command to prevent R from converting all columns of character type (strings) to the <code class="docutils literal notranslate"><span class="pre">factor</span></code> data type (this will create problems with subsequent data wrangling).</p>
</div>
<p>★ Now load the Metadata:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyMetaData</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;../data/PoundHillMetaData.csv&quot;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>  <span class="n">sep</span><span class="o">=</span><span class="s">&quot;;&quot;</span><span class="p">)</span>
<span class="nf">class</span><span class="p">(</span><span class="n">MyMetaData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'data.frame'</div></div>
</div>
<ul class="simple">
<li><p>Here, <code class="docutils literal notranslate"><span class="pre">header</span> <span class="pre">=TRUE</span></code> because we do have metadata headers (<code class="docutils literal notranslate"><span class="pre">FieldName</span></code> and<code class="docutils literal notranslate"><span class="pre">Description</span></code>), and</p></li>
<li><p>We have used semicolon (<code class="docutils literal notranslate"><span class="pre">;</span></code>) as delimiter because there are commas in one of the field descriptions.</p></li>
<li><p>We have avoided spaces in the columns headers (so “FieldName” instead of “Field Name”) — please avoid spaces in field or column names because R will replace each space in a column header with a dot, which may be confusing.</p></li>
</ul>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><strong>The text encoding of your data file:</strong> If you have a string (<code class="docutils literal notranslate"><span class="pre">character</span></code> data type) stored as a variable in your R workspace, or a file containing strings, the computer has to know what <a class="reference external" href="https://en.wikipedia.org/wiki/Character_encoding">character  encoding</a> it is in or it cannot interpret or display it to you correctly. Usually, the encoding will be UTF-8 or ASCII, which is easily handled by most computer languages. Sometimes you may run into (unexpected) bugs when importing and running scripts in <code class="docutils literal notranslate"><span class="pre">R</span></code> because your file has a non-standard text encoding. You will need to specify the encoding in that case, using the <code class="docutils literal notranslate"><span class="pre">encoding</span></code> argument of <code class="docutils literal notranslate"><span class="pre">read.csv()</span></code> and <code class="docutils literal notranslate"><span class="pre">read.table()</span></code>. You can check the encoding of a file by using <code class="docutils literal notranslate"><span class="pre">find</span></code> in Linux/Mac. Try in your UNIX terminal:</p>
<p><code class="docutils literal notranslate"><span class="pre">file</span> <span class="pre">-i</span> <span class="pre">../data/PoundHillData.csv</span></code></p>
<p>or, check encoding of all files in the <code class="docutils literal notranslate"><span class="pre">Data</span></code> directory:</p>
<p><code class="docutils literal notranslate"><span class="pre">file</span> <span class="pre">-i</span> <span class="pre">../data/*.csv</span></code></p>
<p>Use <code class="docutils literal notranslate"><span class="pre">file</span> <span class="pre">-I</span></code> instead of <code class="docutils literal notranslate"><span class="pre">file</span> <span class="pre">-i</span></code> in a Mac terminal.</p>
</div>
<p>Now check out what the data look like:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">MyData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A matrix: 6 × 60 of type chr</caption>
<thead>
	<tr><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th><th scope=col>V6</th><th scope=col>V7</th><th scope=col>V8</th><th scope=col>V9</th><th scope=col>V10</th><th scope=col>⋯</th><th scope=col>V51</th><th scope=col>V52</th><th scope=col>V53</th><th scope=col>V54</th><th scope=col>V55</th><th scope=col>V56</th><th scope=col>V57</th><th scope=col>V58</th><th scope=col>V59</th><th scope=col>V60</th></tr>
</thead>
<tbody>
	<tr><td>Cultivation         </td><td>october</td><td>october</td><td>october</td><td>october</td><td>october</td><td>may</td><td>may</td><td>may</td><td>may</td><td>⋯</td><td>march</td><td>march</td><td>may</td><td>may</td><td>may</td><td>october</td><td>october</td><td>october</td><td>october</td><td>october</td></tr>
	<tr><td>Block               </td><td>a      </td><td>a      </td><td>a      </td><td>a      </td><td>a      </td><td>a  </td><td>a  </td><td>a  </td><td>a  </td><td>⋯</td><td>d    </td><td>d    </td><td>d  </td><td>d  </td><td>d  </td><td>d      </td><td>d      </td><td>d      </td><td>d      </td><td>d      </td></tr>
	<tr><td>Plot                </td><td>1      </td><td>1      </td><td>1      </td><td>1      </td><td>1      </td><td>2  </td><td>2  </td><td>2  </td><td>2  </td><td>⋯</td><td>10   </td><td>10   </td><td>12 </td><td>12 </td><td>12 </td><td>11     </td><td>11     </td><td>11     </td><td>11     </td><td>11     </td></tr>
	<tr><td>Quadrat             </td><td>Q1     </td><td>Q2     </td><td>Q3     </td><td>Q4     </td><td>Q5     </td><td>Q1 </td><td>Q2 </td><td>Q3 </td><td>Q4 </td><td>⋯</td><td>Q5   </td><td>Q6   </td><td>Q1 </td><td>Q2 </td><td>Q4 </td><td>Q1     </td><td>Q2     </td><td>Q3     </td><td>Q4     </td><td>Q5     </td></tr>
	<tr><td>Achillea millefolium</td><td>4      </td><td>8      </td><td>3      </td><td>20     </td><td>6      </td><td>4  </td><td>   </td><td>   </td><td>   </td><td>⋯</td><td>4    </td><td>10   </td><td>   </td><td>   </td><td>   </td><td>       </td><td>       </td><td>12     </td><td>6      </td><td>5      </td></tr>
	<tr><td>Agrostis gigantea   </td><td>       </td><td>       </td><td>       </td><td>       </td><td>15     </td><td>   </td><td>   </td><td>   </td><td>   </td><td>⋯</td><td>19   </td><td>80   </td><td>33 </td><td>145</td><td>45 </td><td>62     </td><td>25     </td><td>57     </td><td>113    </td><td>12     </td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Note that column names V1-V60 were generated automatically by R when you imported the data</p>
<p>In RStudio you can also do <code class="docutils literal notranslate"><span class="pre">view(MyData)</span></code> at the R prompt or any other code editor, <code class="docutils literal notranslate"><span class="pre">fix(MyData)</span></code>. We won’t do anything with the metadata file in this session except inspect the information it contains.</p>
<p>Now let’s implement some of the key data wrangling principles listed above on this dataset. You should keep a record of what you are doing to the data using a R script file. This is illustrated in the <code class="docutils literal notranslate"><span class="pre">DataWrang.R</span></code> script file.</p>
<p>★ Copy the script <code class="docutils literal notranslate"><span class="pre">DataWrang.R</span></code> into your own R training directory’s <code class="docutils literal notranslate"><span class="pre">code</span></code> directory and open it.</p>
<p>Go through the script carefully line by line, and make sure you understand what’s going on. Read the comments — add to them if you want.</p>
<p>OK, so onward with our data wrangling.</p>
</div>
<div class="section" id="keep-a-metadata-file-for-each-unique-dataset">
<h3>Keep a metadata file for each unique dataset<a class="headerlink" href="#keep-a-metadata-file-for-each-unique-dataset" title="Permalink to this headline">¶</a></h3>
<p>Data wrangling really begins immediately after data collection. You may collect data of different kinds (e.g., diversity, biomass, tree girth), etc. Keep the original spreadsheet well documented using a “metadata” file that describes the data (you would hopefully have written the first version of this even before you started collecting the data!). The minimum information needed to make a metadata file useful is a description of each of the <em>fields</em> — the column or row headers under which the information is stored in your data/spreadsheet.</p>
<p>Have a look at the metadata file for the Pound Hill dataset:</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><strong>Boolean arguments in R</strong>: In <code class="docutils literal notranslate"><span class="pre">R</span></code>, you can use <code class="docutils literal notranslate"><span class="pre">F</span></code> and <code class="docutils literal notranslate"><span class="pre">T</span></code> for boolean <code class="docutils literal notranslate"><span class="pre">FALSE</span></code> and <code class="docutils literal notranslate"><span class="pre">TRUE</span></code> respectively. To see this, type</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">a</span> <span class="o">&lt;-</span> <span class="bp">T</span>
</pre></div>
</div>
</div>
<p>in the R commandline, and then see what R returns when you type <code class="docutils literal notranslate"><span class="pre">a</span></code>. Using <code class="docutils literal notranslate"><span class="pre">F</span></code> and <code class="docutils literal notranslate"><span class="pre">T</span></code> for boolean <code class="docutils literal notranslate"><span class="pre">FALSE</span></code> and <code class="docutils literal notranslate"><span class="pre">TRUE</span></code> respectively is not necessarily good practice, but be aware that this option exists.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyMetaData</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 5 × 2</caption>
<thead>
	<tr><th scope=col>FieldName</th><th scope=col>Description</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><td>Cultivation </td><td> Cultivation treatments applied in three months: october, may, march</td></tr>
	<tr><td>Block       </td><td> Treatment blocks ids: a-d                                          </td></tr>
	<tr><td>Plot        </td><td> Plot ids under each treatment : 1-12                               </td></tr>
	<tr><td>Quadrat     </td><td> Sampling quadrats (25x50 cm each) per plot: Q1--Q6                 </td></tr>
	<tr><td>SpeciesCount</td><td> Number of individuals of species (count) per quadrat               </td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Ideally, you would also like to add more information about the data, such as the measurement units of each type of observation. These data include just one type of observation: Number of individuals of species per sample (plot), which is a count (integer, or <code class="docutils literal notranslate"><span class="pre">int</span></code> data class).</p>
</div>
<div class="section" id="minimize-modifying-raw-data-by-hand">
<h3>Minimize modifying raw data by hand<a class="headerlink" href="#minimize-modifying-raw-data-by-hand" title="Permalink to this headline">¶</a></h3>
<p>When the dataset is large (e.g., 1000’s of rows), cleaning and exploring it can get tricky, and you are very likely to make many mistakes. You should record all the steps you used to process it with an R script rather than risking a manual and basically <em>irreproducible</em> processing. Most importantly, <em>avoid or minimize editing your raw data
file</em>—make a copy (with a meaningful tag in the file name to indicate the date and author) before making hand edits.</p>
<p>All blank cells in the data are <em>true absences</em>, in the sense that species was actually not present in that quadrat. So we can replace those blanks with zeros:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyData</span><span class="p">[</span><span class="n">MyData</span> <span class="o">==</span> <span class="s">&quot;&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="m">0</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="convert-wide-format-data-to-long-format-using-scripts">
<h3>Convert wide format data to long format using scripts<a class="headerlink" href="#convert-wide-format-data-to-long-format-using-scripts" title="Permalink to this headline">¶</a></h3>
<p>One typically records data in the field or experiments using a “wide” format, where a subject’s (e.g., habitat, plot, treatment, species etc) repeated responses or observations (e.g., species count, biomass, etc) will be in a single row, and each response in a separate column. The raw Pound Hill data were recorded in this way. However, the wide format is not ideal for data analysis — instead you need the data in a “long” format, where each row is one observation point per subject. So each subject will have data in multiple rows. Any measures/variables/observations that don’t change across the subjects will have the same value in all the rows. For humans, the wide format is generally more intuitive for recording (e.g., in field data sheets) data. However, for data inspection and analysis, the long format is preferable for two main reasons:</p>
<ul class="simple">
<li><p>If you have many response and/or treatment variables, it is hard to inspect the data values in a wide-form version of the dataset. In the case of the pound hill dataset, the response variable is species, and treatment variable is cultivation month (with sequentially nested replicates—block, plot, quadrat— within it), and the data values are the number (count) of individuals of each species per quadrat. As you can see, there are a a large number of columns (60 to be exact), with columns V2-V60 containing different treatment combinations. This makes it hard to visually see problems with the data values. You would have to look across all these columns to see any issues, or if you wanted to run a single command on all the data values (e.g., <code class="docutils literal notranslate"><span class="pre">is.integer()</span></code> to check if they are all integers, as you would expect), it would be harder to do so or interpret the output if all the species counts were in a single column.</p></li>
<li><p>Long-form datasets are typically required for statistical analysis and visualization packages or commands in R (or Python, for that matter). For example, if you wanted to fit a linear model using R’s <code class="docutils literal notranslate"><span class="pre">lm()</span></code> command, with treatment (cultivation month) as the independent variable, you would need to structure your data in long form. Similarly, and if you wanted to plot histograms of species numbers by treatment using ggplot (coming up), you would also need these data in long format.</p></li>
</ul>
<p><em>OK, so let’s go from wide to long format already!</em></p>
<p>You can switch between wide and long formats using <code class="docutils literal notranslate"><span class="pre">melt()</span></code> and <code class="docutils literal notranslate"><span class="pre">dcast()</span></code> from the <code class="docutils literal notranslate"><span class="pre">reshape2</span></code> package, as illustrated in the script <code class="docutils literal notranslate"><span class="pre">DataWrang.R</span></code> available at <a class="reference external" href="https://github.com/mhasoba/TheMulQuaBio/tree/master/content/code">TheMulQuaBio repository</a>. But first, let’s transpose the data, because for a long format, the (nested) treatments variables should be in rows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyData</span> <span class="o">&lt;-</span> <span class="nf">t</span><span class="p">(</span><span class="n">MyData</span><span class="p">)</span> 
<span class="nf">head</span><span class="p">(</span><span class="n">MyData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A matrix: 6 × 45 of type chr</caption>
<tbody>
	<tr><th scope=row>V1</th><td>Cultivation</td><td>Block</td><td>Plot</td><td>Quadrat</td><td>Achillea millefolium</td><td>Agrostis gigantea</td><td>Anagallis arvensis</td><td>Anchusa arvensis</td><td>Anisantha sterilis</td><td>Aphanes australis</td><td>⋯</td><td>Semecio jacobaea</td><td>Sonchus asper</td><td>Spergula arvensis</td><td>Stellaria graminea</td><td>Taraxacum officinale</td><td>Tripleurospermum inodorum </td><td>Veronica arvensis</td><td>Veronica persica</td><td>Viola arvensis</td><td>Vulpia myuros </td></tr>
	<tr><th scope=row>V2</th><td>october    </td><td>a    </td><td>1   </td><td>Q1     </td><td>4                   </td><td>0                </td><td>0                 </td><td>0               </td><td>0                 </td><td>0                </td><td>⋯</td><td>0               </td><td>0            </td><td>0                </td><td>8                 </td><td>0                   </td><td>11                        </td><td>0                </td><td>1               </td><td>0             </td><td>0             </td></tr>
	<tr><th scope=row>V3</th><td>october    </td><td>a    </td><td>1   </td><td>Q2     </td><td>8                   </td><td>0                </td><td>0                 </td><td>3               </td><td>0                 </td><td>0                </td><td>⋯</td><td>0               </td><td>0            </td><td>0                </td><td>4                 </td><td>0                   </td><td>1                         </td><td>0                </td><td>0               </td><td>3             </td><td>0             </td></tr>
	<tr><th scope=row>V4</th><td>october    </td><td>a    </td><td>1   </td><td>Q3     </td><td>3                   </td><td>0                </td><td>0                 </td><td>1               </td><td>0                 </td><td>0                </td><td>⋯</td><td>0               </td><td>0            </td><td>0                </td><td>1                 </td><td>0                   </td><td>0                         </td><td>0                </td><td>0               </td><td>1             </td><td>0             </td></tr>
	<tr><th scope=row>V5</th><td>october    </td><td>a    </td><td>1   </td><td>Q4     </td><td>20                  </td><td>0                </td><td>0                 </td><td>1               </td><td>0                 </td><td>0                </td><td>⋯</td><td>0               </td><td>0            </td><td>0                </td><td>0                 </td><td>0                   </td><td>0                         </td><td>0                </td><td>0               </td><td>1             </td><td>0             </td></tr>
	<tr><th scope=row>V6</th><td>october    </td><td>a    </td><td>1   </td><td>Q5     </td><td>6                   </td><td>15               </td><td>0                 </td><td>0               </td><td>5                 </td><td>0                </td><td>⋯</td><td>0               </td><td>0            </td><td>0                </td><td>1                 </td><td>10                  </td><td>0                         </td><td>0                </td><td>0               </td><td>0             </td><td>7             </td></tr>
</tbody>
</table>
</div></div>
</div>
<p>At this point, you should note that the first row in the data <code class="docutils literal notranslate"><span class="pre">matrix</span></code> actually contains the column headers, but these are not encoded by R as being column headers. As far as R is concerned, the first row is just another data row. You can check this:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">colnames</span><span class="p">(</span><span class="n">MyData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>NULL
</pre></div>
</div>
</div>
</div>
<p>We need to fix this. Also, we would like to create a dataframe now, as this is needed by the R commands we will be using, and ultimately, any visualization and analysis we will be doing. To this end, let’s first create a temporary dataframe with just the data, without the column names:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">TempData</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">MyData</span><span class="p">[</span><span class="m">-1</span><span class="p">,],</span><span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">TempData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 6 × 45</caption>
<thead>
	<tr><th></th><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th><th scope=col>V6</th><th scope=col>V7</th><th scope=col>V8</th><th scope=col>V9</th><th scope=col>V10</th><th scope=col>⋯</th><th scope=col>V36</th><th scope=col>V37</th><th scope=col>V38</th><th scope=col>V39</th><th scope=col>V40</th><th scope=col>V41</th><th scope=col>V42</th><th scope=col>V43</th><th scope=col>V44</th><th scope=col>V45</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>V2</th><td>october</td><td>a</td><td>1</td><td>Q1</td><td>4 </td><td>0 </td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>8</td><td>0 </td><td>11</td><td>0</td><td>1</td><td>0</td><td>0</td></tr>
	<tr><th scope=row>V3</th><td>october</td><td>a</td><td>1</td><td>Q2</td><td>8 </td><td>0 </td><td>0</td><td>3</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>4</td><td>0 </td><td>1 </td><td>0</td><td>0</td><td>3</td><td>0</td></tr>
	<tr><th scope=row>V4</th><td>october</td><td>a</td><td>1</td><td>Q3</td><td>3 </td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0 </td><td>0 </td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><th scope=row>V5</th><td>october</td><td>a</td><td>1</td><td>Q4</td><td>20</td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0 </td><td>0 </td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><th scope=row>V6</th><td>october</td><td>a</td><td>1</td><td>Q5</td><td>6 </td><td>15</td><td>0</td><td>0</td><td>5</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>1</td><td>10</td><td>0 </td><td>0</td><td>0</td><td>0</td><td>7</td></tr>
	<tr><th scope=row>V7</th><td>may    </td><td>a</td><td>2</td><td>Q1</td><td>4 </td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>2</td><td>0</td><td>0 </td><td>0 </td><td>0</td><td>2</td><td>0</td><td>0</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">stringsAsFactors</span> <span class="pre">=</span> <span class="pre">F</span></code> is important here because we don’t want R to convert columns to the factor class without supervision. This might be a good idea in may cases, but let’s just do it manually later. This will also force you to think carefully about what data each of you columns contains.</p>
<p>Now let’s assign the original column names to the temporary dataset:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">colnames</span><span class="p">(</span><span class="n">TempData</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">MyData</span><span class="p">[</span><span class="m">1</span><span class="p">,]</span> <span class="c1"># assign column names from original data</span>
<span class="nf">head</span><span class="p">(</span><span class="n">TempData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 6 × 45</caption>
<thead>
	<tr><th></th><th scope=col>Cultivation</th><th scope=col>Block</th><th scope=col>Plot</th><th scope=col>Quadrat</th><th scope=col>Achillea millefolium</th><th scope=col>Agrostis gigantea</th><th scope=col>Anagallis arvensis</th><th scope=col>Anchusa arvensis</th><th scope=col>Anisantha sterilis</th><th scope=col>Aphanes australis</th><th scope=col>⋯</th><th scope=col>Semecio jacobaea</th><th scope=col>Sonchus asper</th><th scope=col>Spergula arvensis</th><th scope=col>Stellaria graminea</th><th scope=col>Taraxacum officinale</th><th scope=col>Tripleurospermum inodorum </th><th scope=col>Veronica arvensis</th><th scope=col>Veronica persica</th><th scope=col>Viola arvensis</th><th scope=col>Vulpia myuros </th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>V2</th><td>october</td><td>a</td><td>1</td><td>Q1</td><td>4 </td><td>0 </td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>8</td><td>0 </td><td>11</td><td>0</td><td>1</td><td>0</td><td>0</td></tr>
	<tr><th scope=row>V3</th><td>october</td><td>a</td><td>1</td><td>Q2</td><td>8 </td><td>0 </td><td>0</td><td>3</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>4</td><td>0 </td><td>1 </td><td>0</td><td>0</td><td>3</td><td>0</td></tr>
	<tr><th scope=row>V4</th><td>october</td><td>a</td><td>1</td><td>Q3</td><td>3 </td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0 </td><td>0 </td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><th scope=row>V5</th><td>october</td><td>a</td><td>1</td><td>Q4</td><td>20</td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0 </td><td>0 </td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><th scope=row>V6</th><td>october</td><td>a</td><td>1</td><td>Q5</td><td>6 </td><td>15</td><td>0</td><td>0</td><td>5</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>1</td><td>10</td><td>0 </td><td>0</td><td>0</td><td>0</td><td>7</td></tr>
	<tr><th scope=row>V7</th><td>may    </td><td>a</td><td>2</td><td>Q1</td><td>4 </td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>2</td><td>0</td><td>0 </td><td>0 </td><td>0</td><td>2</td><td>0</td><td>0</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>The row names still remain, but they are not really bothering us, so we can ignore them. But if you do want to get rid of them, you can:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">rownames</span><span class="p">(</span><span class="n">TempData</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="kc">NULL</span>
<span class="nf">head</span><span class="p">(</span><span class="n">TempData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 6 × 45</caption>
<thead>
	<tr><th></th><th scope=col>Cultivation</th><th scope=col>Block</th><th scope=col>Plot</th><th scope=col>Quadrat</th><th scope=col>Achillea millefolium</th><th scope=col>Agrostis gigantea</th><th scope=col>Anagallis arvensis</th><th scope=col>Anchusa arvensis</th><th scope=col>Anisantha sterilis</th><th scope=col>Aphanes australis</th><th scope=col>⋯</th><th scope=col>Semecio jacobaea</th><th scope=col>Sonchus asper</th><th scope=col>Spergula arvensis</th><th scope=col>Stellaria graminea</th><th scope=col>Taraxacum officinale</th><th scope=col>Tripleurospermum inodorum </th><th scope=col>Veronica arvensis</th><th scope=col>Veronica persica</th><th scope=col>Viola arvensis</th><th scope=col>Vulpia myuros </th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>october</td><td>a</td><td>1</td><td>Q1</td><td>4 </td><td>0 </td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>8</td><td>0 </td><td>11</td><td>0</td><td>1</td><td>0</td><td>0</td></tr>
	<tr><th scope=row>2</th><td>october</td><td>a</td><td>1</td><td>Q2</td><td>8 </td><td>0 </td><td>0</td><td>3</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>4</td><td>0 </td><td>1 </td><td>0</td><td>0</td><td>3</td><td>0</td></tr>
	<tr><th scope=row>3</th><td>october</td><td>a</td><td>1</td><td>Q3</td><td>3 </td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0 </td><td>0 </td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><th scope=row>4</th><td>october</td><td>a</td><td>1</td><td>Q4</td><td>20</td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0 </td><td>0 </td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><th scope=row>5</th><td>october</td><td>a</td><td>1</td><td>Q5</td><td>6 </td><td>15</td><td>0</td><td>0</td><td>5</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>1</td><td>10</td><td>0 </td><td>0</td><td>0</td><td>0</td><td>7</td></tr>
	<tr><th scope=row>6</th><td>may    </td><td>a</td><td>2</td><td>Q1</td><td>4 </td><td>0 </td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>2</td><td>0</td><td>0 </td><td>0 </td><td>0</td><td>2</td><td>0</td><td>0</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Finally, let’s convert the data to long format. For this, we need the <code class="docutils literal notranslate"><span class="pre">reshape2</span></code> package:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Both <code class="docutils literal notranslate"><span class="pre">library()</span></code> and <code class="docutils literal notranslate"><span class="pre">require()</span></code> are commands/functions to load packages. The difference is that <code class="docutils literal notranslate"><span class="pre">require()</span></code> is designed for use inside other functions, so it returns <code class="docutils literal notranslate"><span class="pre">FALSE</span></code> and gives a warning, whereas<code class="docutils literal notranslate"><span class="pre">library()</span></code> returns an error by default if the package does not exist.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">require</span><span class="p">(</span><span class="n">reshape2</span><span class="p">)</span><span class="c1"># load the reshape2 package</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading required package: reshape2
</pre></div>
</div>
</div>
</div>
<p>Check out the help for the <code class="docutils literal notranslate"><span class="pre">melt</span></code> command of <code class="docutils literal notranslate"><span class="pre">reshape2</span></code>: <code class="docutils literal notranslate"><span class="pre">?melt</span></code>.</p>
<p>OK finally, let’s wrangle this dataset into submission!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyWrangledData</span> <span class="o">&lt;-</span> <span class="nf">melt</span><span class="p">(</span><span class="n">TempData</span><span class="p">,</span> <span class="n">id</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Cultivation&quot;</span><span class="p">,</span> <span class="s">&quot;Block&quot;</span><span class="p">,</span> <span class="s">&quot;Plot&quot;</span><span class="p">,</span> <span class="s">&quot;Quadrat&quot;</span><span class="p">),</span> <span class="n">variable.name</span> <span class="o">=</span> <span class="s">&quot;Species&quot;</span><span class="p">,</span> <span class="n">value.name</span> <span class="o">=</span> <span class="s">&quot;Count&quot;</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">);</span> <span class="nf">tail</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 6 × 6</caption>
<thead>
	<tr><th></th><th scope=col>Cultivation</th><th scope=col>Block</th><th scope=col>Plot</th><th scope=col>Quadrat</th><th scope=col>Species</th><th scope=col>Count</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>october</td><td>a</td><td>1</td><td>Q1</td><td>Achillea millefolium</td><td>4 </td></tr>
	<tr><th scope=row>2</th><td>october</td><td>a</td><td>1</td><td>Q2</td><td>Achillea millefolium</td><td>8 </td></tr>
	<tr><th scope=row>3</th><td>october</td><td>a</td><td>1</td><td>Q3</td><td>Achillea millefolium</td><td>3 </td></tr>
	<tr><th scope=row>4</th><td>october</td><td>a</td><td>1</td><td>Q4</td><td>Achillea millefolium</td><td>20</td></tr>
	<tr><th scope=row>5</th><td>october</td><td>a</td><td>1</td><td>Q5</td><td>Achillea millefolium</td><td>6 </td></tr>
	<tr><th scope=row>6</th><td>may    </td><td>a</td><td>2</td><td>Q1</td><td>Achillea millefolium</td><td>4 </td></tr>
</tbody>
</table>
</div><div class="output text_html"><table>
<caption>A data.frame: 6 × 6</caption>
<thead>
	<tr><th></th><th scope=col>Cultivation</th><th scope=col>Block</th><th scope=col>Plot</th><th scope=col>Quadrat</th><th scope=col>Species</th><th scope=col>Count</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>2414</th><td>may    </td><td>d</td><td>12</td><td>Q4</td><td>Vulpia myuros </td><td>0</td></tr>
	<tr><th scope=row>2415</th><td>october</td><td>d</td><td>11</td><td>Q1</td><td>Vulpia myuros </td><td>0</td></tr>
	<tr><th scope=row>2416</th><td>october</td><td>d</td><td>11</td><td>Q2</td><td>Vulpia myuros </td><td>0</td></tr>
	<tr><th scope=row>2417</th><td>october</td><td>d</td><td>11</td><td>Q3</td><td>Vulpia myuros </td><td>0</td></tr>
	<tr><th scope=row>2418</th><td>october</td><td>d</td><td>11</td><td>Q4</td><td>Vulpia myuros </td><td>0</td></tr>
	<tr><th scope=row>2419</th><td>october</td><td>d</td><td>11</td><td>Q5</td><td>Vulpia myuros </td><td>0</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>This is the long format we have been extolling! It looks much simpler, with all the count data in a single, long column with each count value associated with some “attributes”. It is now clear that the fundamental unit of analysis is a species’ count in a quadrat, each nested within a plot, each in turn nested within a block, and finally, each of which is nested within a cultivation month (the main treatment).</p>
<p>We can also now assign the correct data types to each column:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Cultivation&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Cultivation&quot;</span><span class="p">])</span>
<span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Block&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Block&quot;</span><span class="p">])</span>
<span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Plot&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Plot&quot;</span><span class="p">])</span>
<span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Quadrat&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Quadrat&quot;</span><span class="p">])</span>
<span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Count&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="nf">as.integer</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">[,</span> <span class="s">&quot;Count&quot;</span><span class="p">])</span>
<span class="nf">str</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;data.frame&#39;:	2419 obs. of  6 variables:
 $ Cultivation: Factor w/ 3 levels &quot;march&quot;,&quot;may&quot;,..: 3 3 3 3 3 2 2 2 2 2 ...
 $ Block      : Factor w/ 4 levels &quot;a&quot;,&quot;b&quot;,&quot;c&quot;,&quot;d&quot;: 1 1 1 1 1 1 1 1 1 1 ...
 $ Plot       : Factor w/ 12 levels &quot;1&quot;,&quot;10&quot;,&quot;11&quot;,..: 1 1 1 1 1 5 5 5 5 5 ...
 $ Quadrat    : Factor w/ 6 levels &quot;Q1&quot;,&quot;Q2&quot;,&quot;Q3&quot;,..: 1 2 3 4 5 1 2 3 4 5 ...
 $ Species    : Factor w/ 41 levels &quot;Achillea millefolium&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
 $ Count      : int  4 8 3 20 6 4 0 0 0 0 ...
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="on-to-data-exploration">
<h2>On to data exploration<a class="headerlink" href="#on-to-data-exploration" title="Permalink to this headline">¶</a></h2>
<p>Once you have wrangled the Pound Hill data to its long format, you are ready to go! You may want to start by examining and visualizing the basic properties of the data, such as the number of tree species (41) in the dataset, number of quadrats (replicates) per plot and cultivation treatment, etc.</p>
<p>We will learn about visualization next. After that, you can return to this dataset and try out some visual data exploration. For example, a useful visualization would be to make a histogram of abundances of species, grouped by different factors. For example, you can look at distributions of species’ abundances grouped by the fundamental treatment, <code class="docutils literal notranslate"><span class="pre">Cultivation</span></code>.</p>
<div class="section" id="and-then-came-tidyverse">
<h3>And then came <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code><a class="headerlink" href="#and-then-came-tidyverse" title="Permalink to this headline">¶</a></h3>
<p>So if you think this is the end of the options you have for data wrangling in R, think again. There is a whole data science “ecosystem” you can use in R through the <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> package.  This meta-package includes <code class="docutils literal notranslate"><span class="pre">dplyr</span></code>, the next iteration of <code class="docutils literal notranslate"><span class="pre">plyr</span></code> that addresses the speed issues in the latter, and <code class="docutils literal notranslate"><span class="pre">tidyr</span></code>, essentially a nicer wrapper to the <code class="docutils literal notranslate"> <span class="pre">reshape2</span></code> package with additional functions, and <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code> (coming up later).</p>
<p>Assuming you have installed <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> (see the end of the <span class="xref myst">R Chapter</span>), load it:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">require</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This loads a number of packages at one go (and you might get some name-conflict-masking messages; see the note below). You can list these packages:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">tidyverse_packages</span><span class="p">(</span><span class="n">include_self</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> <span class="c1"># the include_self = TRUE means list &quot;tidyverse&quot; as well </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'broom'</li><li>'cli'</li><li>'crayon'</li><li>'dbplyr'</li><li>'dplyr'</li><li>'forcats'</li><li>'ggplot2'</li><li>'haven'</li><li>'hms'</li><li>'httr'</li><li>'jsonlite'</li><li>'lubridate'</li><li>'magrittr'</li><li>'modelr'</li><li>'pillar'</li><li>'purrr'</li><li>'readr'</li><li>'readxl'</li><li>'reprex'</li><li>'rlang'</li><li>'rstudioapi'</li><li>'rvest'</li><li>'stringr'</li><li>'tibble'</li><li>'tidyr'</li><li>'xml2'</li><li>'tidyverse'</li></ol>
</div></div>
</div>
<p>That’s a lot of packages!</p>
<p>The first time you load tidyverse, you will get some feedback from R, including something like the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
</pre></div>
</div>
<p>Depending on which package you are loading, you might get such a warning message, that “so and so” function is being “masked” from another one. Basically, any additional packages you load may have one or more functions with the same names(s) as existing function(s) (which will typically do something completely different) in base R’s packages. This creates name conflicts. Therefore, R will prevent you from using the new function(s) that you have enabled by using the same name of an pre-existing function. It gives priority to the pre-existing one. So, for example, if you call just <code class="docutils literal notranslate"><span class="pre">filter()</span></code>, the command that will be used is the one from <code class="docutils literal notranslate"><span class="pre">stats</span></code>, and not the one from <code class="docutils literal notranslate"><span class="pre">dplyr</span></code>. In this scenario, you can use the new function by using the double colon notation <code class="docutils literal notranslate"><span class="pre">::</span></code> like so: <code class="docutils literal notranslate"><span class="pre">dplyr::filter()</span></code>.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The <code class="docutils literal notranslate"><span class="pre">::</span></code> notation is like the dot notation in Python — it allows you to access a particular function (like a Python module) from a package. So you can also use the command at the end of the <code class="docutils literal notranslate"><span class="pre">::</span></code> (try just <code class="docutils literal notranslate"><span class="pre">as_tibble(MyWrangledData)</span></code>, for example), but it’s just safer to use the <code class="docutils literal notranslate"><span class="pre">::</span></code> because then you can rest assured that you are actually using the function you intended to.</p>
</div>
<p>More of <code class="docutils literal notranslate"><span class="pre">::</span></code> usage coming up below!</p>
<p>OK, let’s try data exploration in tidyverse.</p>
<p>First, let’s convert the dataframe to a “tibble”:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">tibble</span><span class="o">::</span><span class="nf">as_tibble</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">)</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A tibble: 2419 × 6</caption>
<thead>
	<tr><th scope=col>Cultivation</th><th scope=col>Block</th><th scope=col>Plot</th><th scope=col>Quadrat</th><th scope=col>Species</th><th scope=col>Count</th></tr>
	<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
	<tr><td>october</td><td>a</td><td>1</td><td>Q1</td><td>Achillea millefolium</td><td> 4</td></tr>
	<tr><td>october</td><td>a</td><td>1</td><td>Q2</td><td>Achillea millefolium</td><td> 8</td></tr>
	<tr><td>october</td><td>a</td><td>1</td><td>Q3</td><td>Achillea millefolium</td><td> 3</td></tr>
	<tr><td>october</td><td>a</td><td>1</td><td>Q4</td><td>Achillea millefolium</td><td>20</td></tr>
	<tr><td>october</td><td>a</td><td>1</td><td>Q5</td><td>Achillea millefolium</td><td> 6</td></tr>
	<tr><td>may    </td><td>a</td><td>2</td><td>Q1</td><td>Achillea millefolium</td><td> 4</td></tr>
	<tr><td>may    </td><td>a</td><td>2</td><td>Q2</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>may    </td><td>a</td><td>2</td><td>Q3</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>may    </td><td>a</td><td>2</td><td>Q4</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>may    </td><td>a</td><td>2</td><td>Q5</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>march  </td><td>a</td><td>3</td><td>Q1</td><td>Achillea millefolium</td><td> 1</td></tr>
	<tr><td>march  </td><td>a</td><td>3</td><td>Q2</td><td>Achillea millefolium</td><td> 2</td></tr>
	<tr><td>march  </td><td>a</td><td>3</td><td>Q3</td><td>Achillea millefolium</td><td> 5</td></tr>
	<tr><td>march  </td><td>a</td><td>3</td><td>Q4</td><td>Achillea millefolium</td><td>12</td></tr>
	<tr><td>march  </td><td>a</td><td>3</td><td>Q5</td><td>Achillea millefolium</td><td> 2</td></tr>
	<tr><td>march  </td><td>a</td><td>3</td><td>Q6</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>may    </td><td>b</td><td>4</td><td>Q1</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>may    </td><td>b</td><td>4</td><td>Q2</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>may    </td><td>b</td><td>4</td><td>Q3</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>may    </td><td>b</td><td>4</td><td>Q4</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>may    </td><td>b</td><td>4</td><td>Q5</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>march  </td><td>b</td><td>5</td><td>Q1</td><td>Achillea millefolium</td><td> 3</td></tr>
	<tr><td>march  </td><td>b</td><td>5</td><td>Q2</td><td>Achillea millefolium</td><td> 8</td></tr>
	<tr><td>march  </td><td>b</td><td>5</td><td>Q3</td><td>Achillea millefolium</td><td> 3</td></tr>
	<tr><td>march  </td><td>b</td><td>5</td><td>Q4</td><td>Achillea millefolium</td><td> 7</td></tr>
	<tr><td>march  </td><td>b</td><td>5</td><td>Q5</td><td>Achillea millefolium</td><td>20</td></tr>
	<tr><td>october</td><td>b</td><td>6</td><td>Q1</td><td>Achillea millefolium</td><td> 1</td></tr>
	<tr><td>october</td><td>b</td><td>6</td><td>Q2</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>october</td><td>b</td><td>6</td><td>Q3</td><td>Achillea millefolium</td><td> 3</td></tr>
	<tr><td>october</td><td>b</td><td>6</td><td>Q4</td><td>Achillea millefolium</td><td> 1</td></tr>
	<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>
	<tr><td>october</td><td>b</td><td>6 </td><td>Q4</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>c</td><td>7 </td><td>Q1</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>c</td><td>7 </td><td>Q2</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>c</td><td>7 </td><td>Q3</td><td>Vulpia myuros </td><td> 1</td></tr>
	<tr><td>march  </td><td>c</td><td>7 </td><td>Q4</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>c</td><td>7 </td><td>Q5</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q1</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q2</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q3</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q4</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q5</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>may    </td><td>c</td><td>9 </td><td>Q1</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>may    </td><td>c</td><td>9 </td><td>Q2</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>may    </td><td>c</td><td>9 </td><td>Q3</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>may    </td><td>c</td><td>9 </td><td>Q4</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>may    </td><td>c</td><td>9 </td><td>Q5</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q1</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q2</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q3</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q4</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q5</td><td>Vulpia myuros </td><td>25</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q6</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>may    </td><td>d</td><td>12</td><td>Q1</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>may    </td><td>d</td><td>12</td><td>Q2</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>may    </td><td>d</td><td>12</td><td>Q4</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>d</td><td>11</td><td>Q1</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>d</td><td>11</td><td>Q2</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>d</td><td>11</td><td>Q3</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>d</td><td>11</td><td>Q4</td><td>Vulpia myuros </td><td> 0</td></tr>
	<tr><td>october</td><td>d</td><td>11</td><td>Q5</td><td>Vulpia myuros </td><td> 0</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>OK, there’s the funky double colon (<code class="docutils literal notranslate"><span class="pre">::</span></code>) notation again…</p>
<p>Also, note that we didn’t have to reassign <code class="docutils literal notranslate"><span class="pre">MyWrangledData</span></code> back to its original name because tibble just does the needful to the data frame in its current place.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A “tibble” in tidyverse is equivalent to R’s traditional <code class="docutils literal notranslate"><span class="pre">data.frame</span></code>. Tibbles <em>are</em> modified data frames, but they make data exploration even easier.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dplyr</span><span class="o">::</span><span class="nf">glimpse</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">)</span> <span class="c1">#like str(), but nicer!</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Rows: 2,419
Columns: 6
$ Cultivation <span class=" -Color -Color-C246">&lt;fct&gt;</span> october, october, october, october, october, may, may, ma…
$ Block       <span class=" -Color -Color-C246">&lt;fct&gt;</span> a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, b, b, b, …
$ Plot        <span class=" -Color -Color-C246">&lt;fct&gt;</span> 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, …
$ Quadrat     <span class=" -Color -Color-C246">&lt;fct&gt;</span> Q1, Q2, Q3, Q4, Q5, Q1, Q2, Q3, Q4, Q5, Q1, Q2, Q3, Q4, Q…
$ Species     <span class=" -Color -Color-C246">&lt;fct&gt;</span> Achillea millefolium, Achillea millefolium, Achillea mill…
$ Count       <span class=" -Color -Color-C246">&lt;int&gt;</span> 4, 8, 3, 20, 6, 4, 0, 0, 0, 0, 1, 2, 5, 12, 2, 0, 0, 0, 0…
</pre></div>
</div>
</div>
</div>
<p>Or this:</p>
<p><code class="docutils literal notranslate"><span class="pre">utils::View(MyWrangledData)</span> <span class="pre">#same</span> <span class="pre">as</span> <span class="pre">fix()</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">,</span> <span class="n">Count</span><span class="o">&gt;</span><span class="m">100</span><span class="p">)</span> <span class="c1">#like subset(), but nicer!</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 21 × 6</caption>
<thead>
	<tr><th scope=col>Cultivation</th><th scope=col>Block</th><th scope=col>Plot</th><th scope=col>Quadrat</th><th scope=col>Species</th><th scope=col>Count</th></tr>
	<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
	<tr><td>october</td><td>b</td><td>6 </td><td>Q1</td><td>Agrostis gigantea </td><td>143</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q1</td><td>Agrostis gigantea </td><td>198</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q5</td><td>Agrostis gigantea </td><td>351</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q1</td><td>Agrostis gigantea </td><td>259</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q4</td><td>Agrostis gigantea </td><td>291</td></tr>
	<tr><td>may    </td><td>d</td><td>12</td><td>Q2</td><td>Agrostis gigantea </td><td>145</td></tr>
	<tr><td>october</td><td>d</td><td>11</td><td>Q4</td><td>Agrostis gigantea </td><td>113</td></tr>
	<tr><td>may    </td><td>c</td><td>9 </td><td>Q2</td><td>Anisantha sterilis</td><td>120</td></tr>
	<tr><td>october</td><td>a</td><td>1 </td><td>Q1</td><td>Holcus mollis     </td><td>159</td></tr>
	<tr><td>october</td><td>a</td><td>1 </td><td>Q3</td><td>Holcus mollis     </td><td>144</td></tr>
	<tr><td>october</td><td>a</td><td>1 </td><td>Q4</td><td>Holcus mollis     </td><td>124</td></tr>
	<tr><td>october</td><td>a</td><td>1 </td><td>Q5</td><td>Holcus mollis     </td><td>115</td></tr>
	<tr><td>may    </td><td>a</td><td>2 </td><td>Q1</td><td>Holcus mollis     </td><td>120</td></tr>
	<tr><td>october</td><td>b</td><td>6 </td><td>Q2</td><td>Holcus mollis     </td><td>129</td></tr>
	<tr><td>october</td><td>b</td><td>6 </td><td>Q3</td><td>Holcus mollis     </td><td>123</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q2</td><td>Holcus mollis     </td><td>256</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q3</td><td>Holcus mollis     </td><td>147</td></tr>
	<tr><td>october</td><td>c</td><td>8 </td><td>Q4</td><td>Holcus mollis     </td><td>128</td></tr>
	<tr><td>october</td><td>d</td><td>11</td><td>Q1</td><td>Holcus mollis     </td><td>104</td></tr>
	<tr><td>october</td><td>d</td><td>11</td><td>Q2</td><td>Holcus mollis     </td><td>131</td></tr>
	<tr><td>march  </td><td>d</td><td>10</td><td>Q3</td><td>Medicago lupulina </td><td>114</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dplyr</span><span class="o">::</span><span class="nf">slice</span><span class="p">(</span><span class="n">MyWrangledData</span><span class="p">,</span> <span class="m">10</span><span class="o">:</span><span class="m">15</span><span class="p">)</span> <span class="c1"># Look at an arbitrary set of data rows</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 6 × 6</caption>
<thead>
	<tr><th scope=col>Cultivation</th><th scope=col>Block</th><th scope=col>Plot</th><th scope=col>Quadrat</th><th scope=col>Species</th><th scope=col>Count</th></tr>
	<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
	<tr><td>may  </td><td>a</td><td>2</td><td>Q5</td><td>Achillea millefolium</td><td> 0</td></tr>
	<tr><td>march</td><td>a</td><td>3</td><td>Q1</td><td>Achillea millefolium</td><td> 1</td></tr>
	<tr><td>march</td><td>a</td><td>3</td><td>Q2</td><td>Achillea millefolium</td><td> 2</td></tr>
	<tr><td>march</td><td>a</td><td>3</td><td>Q3</td><td>Achillea millefolium</td><td> 5</td></tr>
	<tr><td>march</td><td>a</td><td>3</td><td>Q4</td><td>Achillea millefolium</td><td>12</td></tr>
	<tr><td>march</td><td>a</td><td>3</td><td>Q5</td><td>Achillea millefolium</td><td> 2</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Learning to use <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> involves learning some new syntax and a lot of new commands, but if you plan to do a lot of data wrangling and exploration, you will benefit from getting to know them well.</p>
<p>Have a look at some tidyverse-related <a class="reference external" href="https://rstudio.com/resources/cheatsheets/">cheatsheets</a> to explore further.</p>
<p>One additional thing that you would need to get used to in the <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> is piping multiple commands using the <code class="docutils literal notranslate"><span class="pre">%%</span></code> notation, which makes your code ore compact, and in theory, more readable.</p>
</div>
</div>
<div class="section" id="practicals">
<h2>Practicals<a class="headerlink" href="#practicals" title="Permalink to this headline">¶</a></h2>
<div class="section" id="data-wrangling-with-tidyverse">
<h3>Data wrangling with <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code><a class="headerlink" href="#data-wrangling-with-tidyverse" title="Permalink to this headline">¶</a></h3>
<p>We used <code class="docutils literal notranslate"><span class="pre">reshape2</span></code> in <code class="docutils literal notranslate"><span class="pre">DataWrang.R</span></code> for wrangling that dataset. Write a new script called <code class="docutils literal notranslate"><span class="pre">DataWrangTidy.R</span></code> that uses <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> (and specifically, <code class="docutils literal notranslate"><span class="pre">dplyr</span></code> and <code class="docutils literal notranslate"><span class="pre">tidyr</span></code>) for the same data wrangling steps. The best way to do this is to copy <code class="docutils literal notranslate"><span class="pre">DataWrang.R</span></code> and rename it <code class="docutils literal notranslate"><span class="pre">DataWrangTidy.R</span></code>. Then systematically redo the script from start to end, looking for a function in <code class="docutils literal notranslate"><span class="pre">dplyr</span></code> and <code class="docutils literal notranslate"> <span class="pre">tidyr</span></code> that does the same thing in each wrangling step.</p>
<p>For example, to convert from wide to long format, instead of using <code class="docutils literal notranslate"> <span class="pre">melt()</span></code> (or <code class="docutils literal notranslate"><span class="pre">dcast()</span></code>) from the <code class="docutils literal notranslate"><span class="pre">reshape2</span></code> package, you can use <code class="docutils literal notranslate"><span class="pre">gather()</span></code> from <code class="docutils literal notranslate"><span class="pre">tidyr</span></code>.</p>
<p>Don’t forget the <a class="reference external" href="https://rstudio.com/resources/cheatsheets/">cheatsheets</a>, an particular the <a class="reference external" href="https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf">Data Transformation Cheatsheet</a>.</p>
</div>
</div>
<div class="section" id="handling-big-data-in-r">
<h2>Handling Big Data in R<a class="headerlink" href="#handling-big-data-in-r" title="Permalink to this headline">¶</a></h2>
<p>The buzzword ‘Big Data’ basically refers to datasets that have the following properties:</p>
<ol class="simple">
<li><p>A dataset that does not fit into available RAM on one system (say, 2 gigabytes).</p></li>
<li><p>A dataset that has so many rows (when in it’s long format — see above sections) that it <em>significantly</em> slows down your analysis or simulation without vectorization (that is, when looping).</p></li>
</ol>
<p>Both these criteria are programming language-, and computer hardware-dependent, of course. For example, a 32-bit OS can only handle ~2 GB of RAM, so this computer will struggle every time you try to handle a dataset in that size range.</p>
<p>R reads data into RAM all at once when you using the <code class="docutils literal notranslate"><span class="pre">read.table</span></code> (or its wrapper, <code class="docutils literal notranslate"><span class="pre">read.csv()</span></code> — maybe
you have realized by now that <code class="docutils literal notranslate"><span class="pre">read.csv()</span></code> is basically calling <code class="docutils literal notranslate"><span class="pre">read.table()</span></code> with a particular set of options. That is, objects in R live in memory entirely, and big-ish data in RAM will cause R to choke. Python has similar problems, but you can circumvent these to an extent by using <code class="docutils literal notranslate"><span class="pre">numpy</span></code> arrays (see the <span class="xref myst">second Python Chapter</span>).</p>
<p>There are a few options (which you can combine, of course) if you are actually using datasets that are so large:</p>
<ul class="simple">
<li><p>Import large files smartly; e.g., using <code class="docutils literal notranslate"><span class="pre">scan()</span></code> in R, and then create subsets of the data that you need. Also, use the <code class="docutils literal notranslate"><span class="pre">reshape</span></code> or <code class="docutils literal notranslate"><span class="pre">tidyr</span></code> packages to covert your data in the most “square” (so neither too long or too wide) format as possible. Of course, you will need subsets of data in long format for analysis (see sections above).</p></li>
<li><p>Use the <code class="docutils literal notranslate"><span class="pre">bigmemory</span></code> package to load data in the gb range (e.g., use <code class="docutils literal notranslate"><span class="pre">read.big.matrix()</span></code> instead of <code class="docutils literal notranslate"><span class="pre">read.table()</span></code>. This package also has other useful functions, such as <code class="docutils literal notranslate"><span class="pre">foreach()</span></code> instead of <code class="docutils literal notranslate"><span class="pre">for()</span></code> for better memory management.</p></li>
<li><p>Use a 64 bit version of R with enough memory and preferably on UNIX/Linux!</p></li>
<li><p>Vectorize your analyses/simulations to the extent possible (See Python and R chapters).</p></li>
<li><p>Use databases (more on this below).</p></li>
<li><p>Use distributed computing (distribute the analysis/simulation across multiple CPU’s).</p></li>
</ul>
<p>The next subsection superficially covers databases. We will cover memory management in the advanced Python, HPC and C weeks.</p>
<div class="section" id="databases-and-r">
<h3>Databases and R<a class="headerlink" href="#databases-and-r" title="Permalink to this headline">¶</a></h3>
<p>R can be used to link to and extract data from online databases such as PubMed and GenBank, or to manipulate and access your own. Computational Biology datasets are often quite large, and it makes sense to access their data by querying the databases instead of manually downloading them. So also, your own data may be complex and large, in which case you may want to organize and manage those data in a proper relational database.</p>
<p>Practically all the data wrangling principles in the previous sections are a part and parcel of relational databases.</p>
<p>There are many R packages that provide an interface to databases (SQLite, MySQL, Oracle, etc). Check out R packages <a class="reference external" href="http://cran.r-project.org/web/packages/DBI/index.html">DBI</a> and <a class="reference external" href="https://cran.r-project.org/web/packages/RMySQL/index.html">RMySQL</a> .</p>
<p>R can also also be used to access, update and manage databases. In particular <code class="docutils literal notranslate"><span class="pre">SQLite</span></code> allows you to build, manipulate, and access databases easily. Try the script called <code class="docutils literal notranslate"><span class="pre">SQLinR.R</span></code> available in the <code class="docutils literal notranslate"><span class="pre">code</span></code> directory of TheMUlQuaBio. Also have a look at <span class="xref myst">this Appendix</span>.</p>
</div>
</div>
<div class="section" id="data-visualization">
<h2>Data visualization<a class="headerlink" href="#data-visualization" title="Permalink to this headline">¶</a></h2>
<p>Now that you have learned how to wrangle data, let’s learn some key principles of graphics and visualization, and how to implement them in R. You can use R to build a complete, reproducible workflow for data visualization for both  exploratory and publication purposes. We will start with some basic plotting and data exploration. You will then learn to generate publication-quality visualizations using the <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code> package.</p>
<div class="section" id="data-exploration-with-basic-plotting">
<h3>Data exploration with basic plotting<a class="headerlink" href="#data-exploration-with-basic-plotting" title="Permalink to this headline">¶</a></h3>
<p>Before you do any statistical analyses with data, you must clean, explore, and visualize it. And eventually, you want to produce a finished product that presents visualizations of your data and your results clearly and concisely.</p>
<p>Ultimately, at both, the data exploration and the finished product stages, the goal of graphics is to present information such that it provides intuitive ideas. As Edward Tufte says:
<em>“Graphical excellence is that which gives to the viewer the greatest number of ideas in the shortest time with the least ink in the smallest space.”</em></p>
<p>R can produce beautiful graphics without the time-consuming and fiddly methods that you might have used in Excel or equivalent. You should also make it a habit to quickly plot the data for exploratory analysis. So we are going to learn some basic plotting first.</p>
<div class="section" id="basic-plotting-commands-in-r">
<h4>Basic plotting commands in R<a class="headerlink" href="#basic-plotting-commands-in-r" title="Permalink to this headline">¶</a></h4>
<p>Here is a menu of basic R plotting commands (use <code class="docutils literal notranslate"><span class="pre">?commandname</span></code> to learn more about it):</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="text-align:left head"><p>Command</p></th>
<th class="text-align:left head"><p>What it does</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-align:left"><p><code class="docutils literal notranslate"><span class="pre">plot(x,y)</span></code></p></td>
<td class="text-align:left"><p>Scatterplot</p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p><code class="docutils literal notranslate"><span class="pre">plot(y~x)</span></code></p></td>
<td class="text-align:left"><p>Scatterplot with <code class="docutils literal notranslate"><span class="pre">y</span></code> as a response variable</p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p><code class="docutils literal notranslate"><span class="pre">hist(mydata)</span></code></p></td>
<td class="text-align:left"><p>Histogram</p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p><code class="docutils literal notranslate"><span class="pre">barplot(mydata)</span></code></p></td>
<td class="text-align:left"><p>Bar plot</p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p><code class="docutils literal notranslate"><span class="pre">points(y1$\sim$x1)</span></code></p></td>
<td class="text-align:left"><p>Add another series of points</p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p><code class="docutils literal notranslate"><span class="pre">boxplot(y$\sim$x)</span></code></p></td>
<td class="text-align:left"><p>Boxplot</p></td>
</tr>
</tbody>
</table>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><strong>R graphics devices</strong>: In all that follows, you may often end up plotting multiple plots on the same graphics window without intending to do so, because R by default keeps plotting in the most recent plotting window that was opened. You can close a particular graphics window or “device” by using <code class="docutils literal notranslate"><span class="pre">dev.off()</span></code>, and all open devices/windows with <code class="docutils literal notranslate"><span class="pre">graphics.off()</span></code>. By default, <code class="docutils literal notranslate"><span class="pre">dev.off()</span></code> will close the most recent figure device that was opened.</p>
</div>
<p>Note that there are invisible devices as well! For example, if you are printing to pdf (coming up below), the device or graphics window will not be visible on your computer screen.</p>
<p>Now let’s try some simple plotting for data exploration. As a case study, we will use a dataset on Consumer-Resource (e.g., Predator-Prey) body mass ratios taken from the Ecological Archives of the ESA (Barnes <em>et al.</em> 2008, Ecology 89:881).</p>
<p><span class="math notranslate nohighlight">\(\star\)</span> Copy <code class="docutils literal notranslate"><span class="pre">EcolArchives-E089-51-D1.csv</span></code> from the TheMulQuaBio <code class="docutils literal notranslate"><span class="pre">data</span></code> to your own data directory. Then, launch R and read in these data to a data frame (note the relative path):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyDF</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;../data/EcolArchives-E089-51-D1.csv&quot;</span><span class="p">)</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">MyDF</span><span class="p">)</span> <span class="c1">#check the size of the data frame you loaded</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>34931</li><li>15</li></ol>
</div></div>
</div>
<p>Let’s look at what the data contain. Type <code class="docutils literal notranslate"><span class="pre">MyDF</span></code> and hit the TAB key twice in the R commandline.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><strong>Remember Tabbing</strong>: In your smart code editor or IDE such as vscode or RStudio, you will see a drop-down list of all the column headers in a dataframe when you hit TAB once after <code class="docutils literal notranslate"><span class="pre">MyDF$</span></code></p>
</div>
<p>You can also use <code class="docutils literal notranslate"><span class="pre">str()</span></code> and <code class="docutils literal notranslate"><span class="pre">head()</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">str</span><span class="p">(</span><span class="n">MyDF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;data.frame&#39;:	34931 obs. of  15 variables:
 $ Record.number              : int  1 2 3 4 5 6 7 8 9 10 ...
 $ In.refID                   : chr  &quot;ATSH063&quot; &quot;ATSH080&quot; &quot;ATSH089&quot; &quot;ATSH143&quot; ...
 $ IndividualID               : chr  &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
 $ Predator                   : chr  &quot;Rhizoprionodon terraenovae&quot; &quot;Rhizoprionodon terraenovae&quot; &quot;Rhizoprionodon terraenovae&quot; &quot;Rhizoprionodon terraenovae&quot; ...
 $ Predator.common.name       : chr  &quot;Atlantic sharpnose shark&quot; &quot;Atlantic sharpnose shark&quot; &quot;Atlantic sharpnose shark&quot; &quot;Atlantic sharpnose shark&quot; ...
 $ Predator.taxon             : chr  &quot;ectotherm vertebrate&quot; &quot;ectotherm vertebrate&quot; &quot;ectotherm vertebrate&quot; &quot;ectotherm vertebrate&quot; ...
 $ Predator.lifestage         : chr  &quot;adult&quot; &quot;adult&quot; &quot;adult&quot; &quot;adult&quot; ...
 $ Type.of.feeding.interaction: chr  &quot;predacious/piscivorous&quot; &quot;predacious/piscivorous&quot; &quot;predacious/piscivorous&quot; &quot;predacious/piscivorous&quot; ...
 $ Predator.mass              : num  1540 1600 1840 87.6 63.9 79.2 71.2 92.1 79.2 79.2 ...
 $ Prey                       : chr  &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; ...
 $ Prey.common.name           : chr  &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; ...
 $ Prey.taxon                 : chr  &quot;mixed&quot; &quot;mixed&quot; &quot;mixed&quot; &quot;mixed&quot; ...
 $ Prey.mass                  : num  14.3 6.02 11.9 8.12 6.56 5.41 4.45 5.98 6.95 5.6 ...
 $ Prey.mass.unit             : chr  &quot;g&quot; &quot;g&quot; &quot;g&quot; &quot;g&quot; ...
 $ Location                   : chr  &quot;Apalachicola Bay, Florida&quot; &quot;Apalachicola Bay, Florida&quot; &quot;Apalachicola Bay, Florida&quot; &quot;Apalachicola Bay, Florida&quot; ...
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">MyDF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 6 × 15</caption>
<thead>
	<tr><th></th><th scope=col>Record.number</th><th scope=col>In.refID</th><th scope=col>IndividualID</th><th scope=col>Predator</th><th scope=col>Predator.common.name</th><th scope=col>Predator.taxon</th><th scope=col>Predator.lifestage</th><th scope=col>Type.of.feeding.interaction</th><th scope=col>Predator.mass</th><th scope=col>Prey</th><th scope=col>Prey.common.name</th><th scope=col>Prey.taxon</th><th scope=col>Prey.mass</th><th scope=col>Prey.mass.unit</th><th scope=col>Location</th></tr>
	<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>1</td><td>ATSH063</td><td>1</td><td>Rhizoprionodon terraenovae</td><td>Atlantic sharpnose shark</td><td>ectotherm vertebrate</td><td>adult</td><td>predacious/piscivorous</td><td>1540.0</td><td>teleosts/molluscs/crustaceans</td><td>teleosts/molluscs/crustaceans</td><td>mixed</td><td>14.30</td><td>g</td><td>Apalachicola Bay, Florida</td></tr>
	<tr><th scope=row>2</th><td>2</td><td>ATSH080</td><td>2</td><td>Rhizoprionodon terraenovae</td><td>Atlantic sharpnose shark</td><td>ectotherm vertebrate</td><td>adult</td><td>predacious/piscivorous</td><td>1600.0</td><td>teleosts/molluscs/crustaceans</td><td>teleosts/molluscs/crustaceans</td><td>mixed</td><td> 6.02</td><td>g</td><td>Apalachicola Bay, Florida</td></tr>
	<tr><th scope=row>3</th><td>3</td><td>ATSH089</td><td>3</td><td>Rhizoprionodon terraenovae</td><td>Atlantic sharpnose shark</td><td>ectotherm vertebrate</td><td>adult</td><td>predacious/piscivorous</td><td>1840.0</td><td>teleosts/molluscs/crustaceans</td><td>teleosts/molluscs/crustaceans</td><td>mixed</td><td>11.90</td><td>g</td><td>Apalachicola Bay, Florida</td></tr>
	<tr><th scope=row>4</th><td>4</td><td>ATSH143</td><td>4</td><td>Rhizoprionodon terraenovae</td><td>Atlantic sharpnose shark</td><td>ectotherm vertebrate</td><td>adult</td><td>predacious/piscivorous</td><td>  87.6</td><td>teleosts/molluscs/crustaceans</td><td>teleosts/molluscs/crustaceans</td><td>mixed</td><td> 8.12</td><td>g</td><td>Apalachicola Bay, Florida</td></tr>
	<tr><th scope=row>5</th><td>5</td><td>ATSH161</td><td>5</td><td>Rhizoprionodon terraenovae</td><td>Atlantic sharpnose shark</td><td>ectotherm vertebrate</td><td>adult</td><td>predacious/piscivorous</td><td>  63.9</td><td>teleosts/molluscs/crustaceans</td><td>teleosts/molluscs/crustaceans</td><td>mixed</td><td> 6.56</td><td>g</td><td>Apalachicola Bay, Florida</td></tr>
	<tr><th scope=row>6</th><td>6</td><td>ATSH166</td><td>6</td><td>Rhizoprionodon terraenovae</td><td>Atlantic sharpnose shark</td><td>ectotherm vertebrate</td><td>adult</td><td>predacious/piscivorous</td><td>  79.2</td><td>teleosts/molluscs/crustaceans</td><td>teleosts/molluscs/crustaceans</td><td>mixed</td><td> 5.41</td><td>g</td><td>Apalachicola Bay, Florida</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>You might want to try the <code class="docutils literal notranslate"><span class="pre">dplyr</span></code> way to do this data inspection/exploration as well, as you did above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">require</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="n">dplyr</span><span class="o">::</span><span class="nf">glimpse</span><span class="p">(</span><span class="n">MyDF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Rows: 34,931
Columns: 15
$ Record.number               <span class=" -Color -Color-C246">&lt;int&gt;</span> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13…
$ In.refID                    <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;ATSH063&quot;, &quot;ATSH080&quot;, &quot;ATSH089&quot;, &quot;ATSH143…
$ IndividualID                <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;1&quot;, &quot;2&quot;, &quot;3&quot;, &quot;4&quot;, &quot;5&quot;, &quot;6&quot;, &quot;7&quot;, &quot;8&quot;, &quot;…
$ Predator                    <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;Rhizoprionodon terraenovae&quot;, &quot;Rhizoprion…
$ Predator.common.name        <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;Atlantic sharpnose shark&quot;, &quot;Atlantic sha…
$ Predator.taxon              <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;ectotherm vertebrate&quot;, &quot;ectotherm verteb…
$ Predator.lifestage          <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;adult&quot;, &quot;adult&quot;, &quot;adult&quot;, &quot;adult&quot;, &quot;adul…
$ Type.of.feeding.interaction <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;predacious/piscivorous&quot;, &quot;predacious/pis…
$ Predator.mass               <span class=" -Color -Color-C246">&lt;dbl&gt;</span> 1540.0, 1600.0, 1840.0, 87.6, 63.9, 79.2,…
$ Prey                        <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;teleosts/molluscs/crustaceans&quot;, &quot;teleost…
$ Prey.common.name            <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;teleosts/molluscs/crustaceans&quot;, &quot;teleost…
$ Prey.taxon                  <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;mixed&quot;, &quot;mixed&quot;, &quot;mixed&quot;, &quot;mixed&quot;, &quot;mixe…
$ Prey.mass                   <span class=" -Color -Color-C246">&lt;dbl&gt;</span> 14.30, 6.02, 11.90, 8.12, 6.56, 5.41, 4.4…
$ Prey.mass.unit              <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;g&quot;, &quot;g&quot;, &quot;g&quot;, &quot;g&quot;, &quot;g&quot;, &quot;g&quot;, &quot;g&quot;, &quot;g&quot;, &quot;…
$ Location                    <span class=" -Color -Color-C246">&lt;chr&gt;</span> &quot;Apalachicola Bay, Florida&quot;, &quot;Apalachicol…
</pre></div>
</div>
</div>
</div>
<p>As you can see, these data contain predator-prey body size information. This is an interesting dataset because it is huge, and covers a wide range of body sizes of aquatic species involved in consumer-resource interactions — from unicells to whales. Analyzing this dataset should tell us a lot about what sizes of prey predators like to eat.</p>
<img src="./graphics/SeaLion.png" alt="A consumer-resource (predator-prey) interaction waiting to happen" style="width: 500px;"/><p>Let’s change the type of certain columns to <code class="docutils literal notranslate"><span class="pre">factor</span></code> because we will want to use them as grouping variables:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">MyDF</span><span class="o">$</span><span class="n">Type.of.feeding.interaction</span> <span class="o">&lt;-</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Type.of.feeding.interaction</span><span class="p">)</span>
<span class="n">MyDF</span><span class="o">$</span><span class="n">Location</span> <span class="o">&lt;-</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Location</span><span class="p">)</span>
<span class="nf">str</span><span class="p">(</span><span class="n">MyDF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;data.frame&#39;:	34931 obs. of  15 variables:
 $ Record.number              : int  1 2 3 4 5 6 7 8 9 10 ...
 $ In.refID                   : chr  &quot;ATSH063&quot; &quot;ATSH080&quot; &quot;ATSH089&quot; &quot;ATSH143&quot; ...
 $ IndividualID               : chr  &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
 $ Predator                   : chr  &quot;Rhizoprionodon terraenovae&quot; &quot;Rhizoprionodon terraenovae&quot; &quot;Rhizoprionodon terraenovae&quot; &quot;Rhizoprionodon terraenovae&quot; ...
 $ Predator.common.name       : chr  &quot;Atlantic sharpnose shark&quot; &quot;Atlantic sharpnose shark&quot; &quot;Atlantic sharpnose shark&quot; &quot;Atlantic sharpnose shark&quot; ...
 $ Predator.taxon             : chr  &quot;ectotherm vertebrate&quot; &quot;ectotherm vertebrate&quot; &quot;ectotherm vertebrate&quot; &quot;ectotherm vertebrate&quot; ...
 $ Predator.lifestage         : chr  &quot;adult&quot; &quot;adult&quot; &quot;adult&quot; &quot;adult&quot; ...
 $ Type.of.feeding.interaction: Factor w/ 5 levels &quot;insectivorous&quot;,..: 5 5 5 5 5 5 5 5 5 5 ...
 $ Predator.mass              : num  1540 1600 1840 87.6 63.9 79.2 71.2 92.1 79.2 79.2 ...
 $ Prey                       : chr  &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; ...
 $ Prey.common.name           : chr  &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; &quot;teleosts/molluscs/crustaceans&quot; ...
 $ Prey.taxon                 : chr  &quot;mixed&quot; &quot;mixed&quot; &quot;mixed&quot; &quot;mixed&quot; ...
 $ Prey.mass                  : num  14.3 6.02 11.9 8.12 6.56 5.41 4.45 5.98 6.95 5.6 ...
 $ Prey.mass.unit             : chr  &quot;g&quot; &quot;g&quot; &quot;g&quot; &quot;g&quot; ...
 $ Location                   : Factor w/ 25 levels &quot;Andaman Sea (West of South Thailand)&quot;,..: 3 3 3 3 3 3 3 3 3 3 ...
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="scatter-plots">
<h3>Scatter Plots<a class="headerlink" href="#scatter-plots" title="Permalink to this headline">¶</a></h3>
<p>Let’s start by plotting Predator mass vs. Prey mass.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">,</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_62_0.png" src="../_images/1-Data_R_62_0.png" />
</div>
</div>
<p>That doesn’t look very meaningful! Let’s try taking logarithms. <em>Why?</em> - Because body sizes across species tend to be log-normally distributed, with a lot of small species and a few large ones. Taking a log allows you to inspect the body size range in a meaningful (logarithmic) scale and reveals the true relationship. This also illustrates a important point. Just like statistical analyses, the effectiveness of data visualization too depends on the type of distribution of the data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_64_0.png" src="../_images/1-Data_R_64_0.png" />
</div>
</div>
<p>Let’s look at the same using a base-10 log transform:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_66_0.png" src="../_images/1-Data_R_66_0.png" />
</div>
</div>
<p>Using a <code class="docutils literal notranslate"><span class="pre">log10</span></code> transform is often a good idea because then you can see things in terms of “orders of magnitude”, (<span class="math notranslate nohighlight">\(10^1\)</span>, <span class="math notranslate nohighlight">\(10^2\)</span>, <span class="math notranslate nohighlight">\(10^3\)</span>, etc), which makes it easier to determine  what the actual values in the original scale are.</p>
<p>We can change almost any aspect of the resulting graph; let’s change the symbols by specifying the <code class="docutils literal notranslate"><span class="pre">p</span></code>lot <code class="docutils literal notranslate"><span class="pre">ch</span></code>aracters using <code class="docutils literal notranslate"><span class="pre">pch</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">),</span><span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">)</span> <span class="c1"># Change marker</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_68_0.png" src="../_images/1-Data_R_68_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">),</span><span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span> <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;Predator Mass (g)&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Prey Mass (g)&quot;</span><span class="p">)</span> <span class="c1"># Add labels</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_69_0.png" src="../_images/1-Data_R_69_0.png" />
</div>
</div>
<p>A really great summary of basic R graphical parameters can be <a class="reference external" href="https://www.statmethods.net/advgraphs/parameters.html">found here</a>.</p>
</div>
<div class="section" id="histograms">
<h3>Histograms<a class="headerlink" href="#histograms" title="Permalink to this headline">¶</a></h3>
<p>Why did we have to take a logarithm to see the relationship between
predator and prey size? Plotting histograms of the two classes
(predator, prey) should be insightful, as we can then see the “marginal”
distributions of the two variables.</p>
<p>Let’s first plot a histogram of predator body masses:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">hist</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_71_0.png" src="../_images/1-Data_R_71_0.png" />
</div>
</div>
<p>Clearly, the data are heavily right skewed, with small body sized organisms dominating (that’s a general pattern, as mentioned above). Let’s now take a logarithm and see if we can get a better idea of what the distribution of predator sizes looks like:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">hist</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;log10(Predator Mass (g))&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Count&quot;</span><span class="p">)</span> <span class="c1"># include labels</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_73_0.png" src="../_images/1-Data_R_73_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">hist</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span><span class="n">xlab</span><span class="o">=</span><span class="s">&quot;log10(Predator Mass (g))&quot;</span><span class="p">,</span><span class="n">ylab</span><span class="o">=</span><span class="s">&quot;Count&quot;</span><span class="p">,</span> 
    <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;lightblue&quot;</span><span class="p">,</span> <span class="n">border</span> <span class="o">=</span> <span class="s">&quot;pink&quot;</span><span class="p">)</span> <span class="c1"># Change bar and borders colors </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_74_0.png" src="../_images/1-Data_R_74_0.png" />
</div>
</div>
<p>So, taking a log really makes clearer what the distribution of body predator sizes looks like. <em>Try the same with prey body masses.</em></p>
<div class="section" id="exercise">
<h4>Exercise<a class="headerlink" href="#exercise" title="Permalink to this headline">¶</a></h4>
<p>We can do a lot of beautification and fine-tuning of your R plots! As an exercise, try adjusting the histogram bin widths to make them same for the predator and prey, and making the x and y labels larger and in boldface. To get started, look at the help documentation of <code class="docutils literal notranslate"><span class="pre">hist</span></code>.</p>
</div>
</div>
<div class="section" id="subplots">
<h3>Subplots<a class="headerlink" href="#subplots" title="Permalink to this headline">¶</a></h3>
<p>We can also plot both predator and prey body masses in different sub-plots using <code class="docutils literal notranslate"><span class="pre">par</span></code> so that we can compare them visually.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">par</span><span class="p">(</span><span class="n">mfcol</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">))</span> <span class="c1">#initialize multi-paneled plot</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mfg</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">))</span> <span class="c1"># specify which sub-plot to use first </span>
<span class="nf">hist</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span>
    <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;log10(Predator Mass (g))&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Count&quot;</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;lightblue&quot;</span><span class="p">,</span> <span class="n">border</span> <span class="o">=</span> <span class="s">&quot;pink&quot;</span><span class="p">,</span> 
    <span class="n">main</span> <span class="o">=</span> <span class="s">&#39;Predator&#39;</span><span class="p">)</span> <span class="c1"># Add title</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mfg</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">))</span> <span class="c1"># Second sub-plot</span>
<span class="nf">hist</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;log10(Prey Mass (g))&quot;</span><span class="p">,</span><span class="n">ylab</span><span class="o">=</span><span class="s">&quot;Count&quot;</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;lightgreen&quot;</span><span class="p">,</span> <span class="n">border</span> <span class="o">=</span> <span class="s">&quot;pink&quot;</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="s">&#39;prey&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_76_0.png" src="../_images/1-Data_R_76_0.png" />
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">par()</span></code> function can set multiple graphics parameters (not just multi-panel plots), including figure margins, axis labels, and more. Check out the <a class="reference external" href="https://www.rdocumentation.org/packages/graphics/versions/3.5.1/topics/par">help for this function</a>.</p>
<p>Another option for making multi-panel plots is the <code class="docutils literal notranslate"><span class="pre">layout</span></code> function.</p>
</div>
<div class="section" id="overlaying-plots">
<h3>Overlaying plots<a class="headerlink" href="#overlaying-plots" title="Permalink to this headline">¶</a></h3>
<p>Better still, we would like to see if the predator mass and prey mass distributions are similar by overlaying them.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">hist</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="c1"># Predator histogram</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;log10(Body Mass (g))&quot;</span><span class="p">,</span> <span class="n">ylab</span><span class="o">=</span><span class="s">&quot;Count&quot;</span><span class="p">,</span> 
    <span class="n">col</span> <span class="o">=</span> <span class="nf">rgb</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">0.5</span><span class="p">),</span> <span class="c1"># Note &#39;rgb&#39;, fourth value is transparency</span>
    <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Predator-prey size Overlap&quot;</span><span class="p">)</span> 
<span class="nf">hist</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">col</span> <span class="o">=</span> <span class="nf">rgb</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="m">0.5</span><span class="p">),</span> <span class="n">add</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span> <span class="c1"># Plot prey</span>
<span class="nf">legend</span><span class="p">(</span><span class="s">&#39;topleft&#39;</span><span class="p">,</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;Predators&#39;</span><span class="p">,</span><span class="s">&#39;Prey&#39;</span><span class="p">),</span>   <span class="c1"># Add legend</span>
    <span class="n">fill</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="nf">rgb</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">0.5</span><span class="p">),</span> <span class="nf">rgb</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="m">0.5</span><span class="p">)))</span> <span class="c1"># Define legend colors</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_78_0.png" src="../_images/1-Data_R_78_0.png" />
</div>
</div>
<p>Plot annotation with text can be done with either single or double quotes, i.e., ‘Plot Title’ or “Plot Title”, respectively. But it is  generally a good idea to use double quotes because sometimes you would like to use an apostrophe in your title or axis label strings.</p>
<p>★ It would be nicer to have both the plots with the same bin sizes – try to do it</p>
</div>
<div class="section" id="boxplots">
<h3>Boxplots<a class="headerlink" href="#boxplots" title="Permalink to this headline">¶</a></h3>
<p>Now, let’s try plotting boxplots instead of histograms. These are useful for getting a visual summary of the distribution of your data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">boxplot</span><span class="p">(</span><span class="nf">log10</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;Location&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;log10(Predator Mass)&quot;</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Predator mass&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_80_0.png" src="../_images/1-Data_R_80_0.png" />
</div>
</div>
<p>Now let’s see how many locations the data are from:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">boxplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">)</span> <span class="o">~</span> <span class="n">MyDF</span><span class="o">$</span><span class="n">Location</span><span class="p">,</span> <span class="c1"># Why the tilde?</span>
    <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;Location&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Predator Mass&quot;</span><span class="p">,</span>
    <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Predator mass by location&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_82_0.png" src="../_images/1-Data_R_82_0.png" />
</div>
</div>
<p>Note the tilde (<code class="docutils literal notranslate"><span class="pre">~</span></code>). This is to tell R to subdivide or categorize your analysis and plot by the “Factor” location. More on this later.</p>
<p>That’s a lot of locations! You will need an appropriately wide plot to see all the boxplots adequately. Now let’s try boxplots by feeding interaction type:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">boxplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">)</span> <span class="o">~</span> <span class="n">MyDF</span><span class="o">$</span><span class="n">Type.of.feeding.interaction</span><span class="p">,</span>
    <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;Location&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Predator Mass&quot;</span><span class="p">,</span>
    <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Predator mass by feeding interaction type&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_84_0.png" src="../_images/1-Data_R_84_0.png" />
</div>
</div>
</div>
<div class="section" id="combining-plot-types">
<h3>Combining plot types<a class="headerlink" href="#combining-plot-types" title="Permalink to this headline">¶</a></h3>
<p>It would be nice to see both the predator and prey (marginal) distributions as well as the scatterplot for an exploratory analysis. We can do this by adding boxplots of the marginal variables to the scatterplot.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span> <span class="nf">par</span><span class="p">(</span><span class="n">fig</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0.8</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0.8</span><span class="p">))</span> <span class="c1"># specify figure size as proportion</span>
 <span class="nf">plot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;Predator Mass (g)&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Prey Mass (g)&quot;</span><span class="p">)</span> <span class="c1"># Add labels</span>
 <span class="nf">par</span><span class="p">(</span><span class="n">fig</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0.8</span><span class="p">,</span><span class="m">0.4</span><span class="p">,</span><span class="m">1</span><span class="p">),</span> <span class="n">new</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
 <span class="nf">boxplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">horizontal</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">axes</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
 <span class="nf">par</span><span class="p">(</span><span class="n">fig</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0.55</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0.8</span><span class="p">),</span><span class="n">new</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
 <span class="nf">boxplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">axes</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
 <span class="nf">mtext</span><span class="p">(</span><span class="s">&quot;Fancy Predator-prey scatterplot&quot;</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">outer</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">line</span><span class="o">=</span><span class="m">-3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_86_0.png" src="../_images/1-Data_R_86_0.png" />
</div>
</div>
<p>To understand this plotting method, think of the full graph area as going from (0,0) in the lower left corner to (1,1) in the upper right corner. The format of the <code class="docutils literal notranslate"><span class="pre">fig=</span></code> parameter is a numerical vector of the form <code class="docutils literal notranslate"><span class="pre">c(x1,</span> <span class="pre">x2,</span> <span class="pre">y1,</span> <span class="pre">y2)</span></code>, corresponding to <code class="docutils literal notranslate"><span class="pre">c(bottom,</span> <span class="pre">left,</span> <span class="pre">top,</span> <span class="pre">right)</span></code>. First, <code class="docutils literal notranslate"> <span class="pre">par(fig=c(0,0.8,0,0.8))</span></code> sets up the scatterplot going from 0 to 0.8 on the x axis and 0 to 0.8 on the y axis, leaving some area for the boxplots at the top and right. The top boxplot goes from 0 to 0.8 on the x axis and 0.4 to 1 on the y axis. The right hand boxplot goes from 0.55 to 1 on the x axis and 0 to 0.8 on the y axis. You can experiment with these proportions to change the spacings between plots.</p>
<p>This plot is useful, because it shows you what the marginal distributions of the two variables (Predator mass and Prey mass) are.</p>
<p>Below you will learn to use <code class="docutils literal notranslate"><span class="pre">ggplot</span></code> to produce a much more elegant, pretty multi-panel plots.</p>
</div>
<div class="section" id="saving-your-graphics">
<h3>Saving your graphics<a class="headerlink" href="#saving-your-graphics" title="Permalink to this headline">¶</a></h3>
<p>And you can also save the figure in a vector graphics format like a pdf. It is important to learn to do this, because you want to be able to save your plots in good resolution, and want to avoid the manual steps of
clicking on the figure, doing “save as”, etc. So let’s save the figure as a PDF:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">pdf</span><span class="p">(</span><span class="s">&quot;../results/Pred_Prey_Overlay.pdf&quot;</span><span class="p">,</span> <span class="c1"># Open blank pdf page using a relative path</span>
    <span class="m">11.7</span><span class="p">,</span> <span class="m">8.3</span><span class="p">)</span> <span class="c1"># These numbers are page dimensions in inches</span>
<span class="nf">hist</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="c1"># Plot predator histogram (note &#39;rgb&#39;)</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;Body Mass (g)&quot;</span><span class="p">,</span> <span class="n">ylab</span><span class="o">=</span><span class="s">&quot;Count&quot;</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="nf">rgb</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">0.5</span><span class="p">),</span> <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Predator-Prey Size Overlap&quot;</span><span class="p">)</span> 
<span class="nf">hist</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">MyDF</span><span class="o">$</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="c1"># Plot prey weights</span>
    <span class="n">col</span> <span class="o">=</span> <span class="nf">rgb</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="m">0.5</span><span class="p">),</span> 
    <span class="n">add</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>  <span class="c1"># Add to same plot = TRUE</span>
<span class="nf">legend</span><span class="p">(</span><span class="s">&#39;topleft&#39;</span><span class="p">,</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;Predators&#39;</span><span class="p">,</span><span class="s">&#39;Prey&#39;</span><span class="p">),</span> <span class="c1"># Add legend</span>
    <span class="n">fill</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="nf">rgb</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">0.5</span><span class="p">),</span> <span class="nf">rgb</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="m">0.5</span><span class="p">)))</span> 
<span class="nf">graphics.off</span><span class="p">();</span> <span class="c1">#you can also use dev.off() </span>
</pre></div>
</div>
</div>
</div>
<p>Always try to save results in a vector format, which can be scaled up to any size. For more on vector vs raster images/graphics, <a class="reference external" href="https://en.wikipedia.org/wiki/Vector_graphics">see this</a>.</p>
<p>Note that you are saving to the <code class="docutils literal notranslate"><span class="pre">results</span></code> directory now. This is a recommended project organization and workflow: store and retrieve data from a <code class="docutils literal notranslate"> <span class="pre">Data</span></code> directory, keep your code and work from a <code class="docutils literal notranslate"><span class="pre">Code</span></code> directory, and save outputs to a <code class="docutils literal notranslate"><span class="pre">results</span></code> directory.</p>
<p>You can also try other graphic output formats. For example, <code class="docutils literal notranslate"><span class="pre">png()</span></code> (a raster format) instead of <code class="docutils literal notranslate"><span class="pre">pdf()</span></code>. As always, look at the help documentation of each of these commands!</p>
</div>
</div>
<div class="section" id="id1">
<h2>Practicals<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<div class="section" id="body-mass-distributions">
<h3>Body mass distributions<a class="headerlink" href="#body-mass-distributions" title="Permalink to this headline">¶</a></h3>
<p>Write a script that draws and saves three figures, each containing subplots of distributions of predator mass, prey mass, and the size ratio of prey mass over predator mass <em>by feeding interaction type</em>. Use logarithms of masses (or size ratios)for all three plots. In addition, the script should calculate the (log) mean and median predator mass, prey mass and predator-prey size-ratios to a csv file.</p>
<p>Call the script <code class="docutils literal notranslate"><span class="pre">PP_Dists.R</span></code> and save it in the <code class="docutils literal notranslate"><span class="pre">Code</span></code> directory — sourcing or running this script should result in three files called <code class="docutils literal notranslate"><span class="pre">Pred_Subplots.pdf</span></code>, <code class="docutils literal notranslate"> <span class="pre">Prey_Subplots.pdf</span></code>, and <code class="docutils literal notranslate"><span class="pre">SizeRatio_Subplots.pdf</span></code> (the names should be self-explanatory), and a single csv file <code class="docutils literal notranslate"><span class="pre">PP_Results.csv</span></code> containing the mean and median log predator mass, prey mass, and predator-prey size ratio, by feeding type. The csv file should have appropriate headers (e.g., Feeding type, Mean, Median). (Hint: you will have to initialize a new dataframe or matrix in the script to first store the calculations)</p>
<p>The script should be self-sufficient and not need any external inputs. It should import the above predator-prey dataset from the appropriate directory, and save the graphic plots to the appropriate directory (Hint: use relative paths!).</p>
<p>Use the <code class="docutils literal notranslate"><span class="pre">par()</span></code> function to create the subplots. For calculating the body size statistics by feeding type, you can either use the “loopy” way — first obtaining a list of feeding types (look up the <code class="docutils literal notranslate"><span class="pre">unique</span></code> or <code class="docutils literal notranslate"><span class="pre">levels</span></code> functions) and then loop over them, using <code class="docutils literal notranslate"><span class="pre">subset</span></code> to extract the dataset by feeding type at each iteration, or the R-savvy way, by using <code class="docutils literal notranslate"><span class="pre">tapply</span></code> or <code class="docutils literal notranslate"><span class="pre">ddply</span></code> and avoiding looping altogether.</p>
</div>
</div>
<div class="section" id="beautiful-graphics-in-r">
<h2>Beautiful graphics in R<a class="headerlink" href="#beautiful-graphics-in-r" title="Permalink to this headline">¶</a></h2>
<p>R can produce beautiful visualizations, but it typically takes a lot of work to obtain the desired result. This is because the starting point is pretty much a “bare” plot, and adding features commonly required for publication-grade figures (legends, statistics, regressions, sub-plotting etc.) can require a lot of small and painful additional arguments to the plotting commands <em>at the same time</em>, or even additional steps (such as the <a class="reference external" href="#Combining-plot-types">fancy predator-prey scatterplot</a> above).</p>
<p>Moreover, it is very difficult to switch from one representation of the data to another (i.e., from boxplots to scatterplots), or to plot several datasets together. The R package <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code> overcomes these issues, and produces truly high-quality, publication-ready graphics suitable for papers, theses and reports.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>3D plots:</strong> <em>Currently, <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code> cannot be used to create 3D graphs or mosaic plots (but <a class="reference external" href="https://www.tylermw.com/3d-ggplots-with-rayshader">see this</a>). In any case, most of you won’t be needing 3D plots. If you do, there are many ways to do 3D plots using other plotting packages in R. In particular, look up the <code class="docutils literal notranslate"><span class="pre">scatterplot3d</span></code> and <code class="docutils literal notranslate"><span class="pre">plot3D</span></code> packages.</em></p>
</div>
<p><code class="docutils literal notranslate"><span class="pre">ggplot2</span></code> differs from other approaches as it attempts to provide a “grammar” for graphics in which each layer is the equivalent of a verb, subject etc. and a plot is the equivalent of a sentence. All graphs start with a layer showing the data, other layers and attributes/styles are added to modify the plot. Specifically, according to this grammar, a statistical graphic is a “mapping” from data to geometric objects (points, lines, bars; set using <code class="docutils literal notranslate"><span class="pre">geom</span></code>) with aesthetic attributes (colour, shape, size; set using <code class="docutils literal notranslate"><span class="pre">aes</span></code>).</p>
<p>For more on the ideas underlying ggplot, see the book “ggplot2: Elegant Graphics for Data Analysis”, by H. Wickham (in your Reading directory). Also, the <a class="reference external" href="http://ggplot2.tidyverse.org/">ggplot2 website</a> is an excellent resource.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code> is not available on your computer, look up the section on installing packages in the <span class="xref myst">R Chapter</span>.</p>
<p>ggplot can be used in two ways: with <code class="docutils literal notranslate"><span class="pre">qplot</span></code> (for <code class="docutils literal notranslate"><span class="pre">q</span></code>uick <code class="docutils literal notranslate"> <span class="pre">plot</span></code>ting) and <code class="docutils literal notranslate"><span class="pre">ggplot</span></code> for fully customized plotting.</p>
<p><em>Note that <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code> only accepts data in data frames.</em></p>
<div class="section" id="quick-plotting-with-qplot">
<h3>Quick plotting with <code class="docutils literal notranslate"><span class="pre">qplot</span></code><a class="headerlink" href="#quick-plotting-with-qplot" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">qplot</span></code> can be used to quickly produce graphics for exploratory data analysis, and as a base for more complex graphics. It uses syntax that is closer to the standard R plotting commands.</p>
<p>We will use the same predator-prey body size dataset again – you will soon see how much nice the same types of plots you made above look when done with ggplot!</p>
<p>First, load the package:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">require</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="scatterplots">
<h4>Scatterplots<a class="headerlink" href="#scatterplots" title="Permalink to this headline">¶</a></h4>
<p>Let’s start plotting the <code class="docutils literal notranslate"><span class="pre">Predator.mass</span></code> vs <code class="docutils literal notranslate"><span class="pre">Prey.mass</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">,</span> <span class="n">Predator.mass</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_94_0.png" src="../_images/1-Data_R_94_0.png" />
</div>
</div>
<p>As before, let’s take logarithms and plot:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_96_0.png" src="../_images/1-Data_R_96_0.png" />
</div>
</div>
<p>Now, color the points according to the type of feeding interaction:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_98_0.png" src="../_images/1-Data_R_98_0.png" />
</div>
</div>
<p>But the figure’s aspect ratio is not very nice. Let’s change it using the <code class="docutils literal notranslate"><span class="pre">asp</span></code> option:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="n">asp</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_100_0.png" src="../_images/1-Data_R_100_0.png" />
</div>
</div>
<p>The same as above, but changing the shape:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">shape</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="n">asp</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_102_0.png" src="../_images/1-Data_R_102_0.png" />
</div>
</div>
</div>
<div class="section" id="aesthetic-mappings">
<h4>Aesthetic mappings<a class="headerlink" href="#aesthetic-mappings" title="Permalink to this headline">¶</a></h4>
<p>These examples demonstrate a key difference between <code class="docutils literal notranslate"><span class="pre">qplot</span></code> (and indeed, ggplot2’s approach) and the standard <code class="docutils literal notranslate"><span class="pre">plot</span></code> command: When you want to assign colours, sizes or shapes to the points on your plot, using the <code class="docutils literal notranslate"><span class="pre">plot</span></code> command, it’s your responsibility to convert (i.e., “map”) a categorical variable in your data (e.g., type of feeding interaction in the above case) onto colors (or shapes) that <code class="docutils literal notranslate"><span class="pre">plot</span></code> knows how to use (e.g., by specifying “red”, “blue”, “green”, etc).</p>
<p>ggplot does this mapping for you automatically, and also provides a legend! This makes it really easy to quickly include additional data (e.g., if a new feeding interaction type was added to the data) on the plot.</p>
<p>Instead of using ggplot’s automatic mapping, if you want to manually set a color or a shape, you have to use <code class="docutils literal notranslate"><span class="pre">I()</span></code> (meaning “Identity”). To see this in practice, try the following:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> 
    <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;red&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_104_0.png" src="../_images/1-Data_R_104_0.png" />
</div>
</div>
<p>You chose red, but ggplot used mapping to convert it to a particular
shade of red. To set it manually to the real red, do this:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">colour</span> <span class="o">=</span> <span class="nf">I</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_106_0.png" src="../_images/1-Data_R_106_0.png" />
</div>
</div>
<p>Similarly, for point size, compare these two:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span> <span class="c1">#with ggplot size mapping</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_108_0.png" src="../_images/1-Data_R_108_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span>  <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="nf">I</span><span class="p">(</span><span class="m">3</span><span class="p">))</span> <span class="c1">#no mapping</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_109_0.png" src="../_images/1-Data_R_109_0.png" />
</div>
</div>
<p>But for shape, ggplot doesn’t have a continuous mapping because shapes are a discrete variable.So</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">shape</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
</pre></div>
</div>
<p>will give an error.</p>
<p>Instead, try this:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span> <span class="nf">I</span><span class="p">(</span><span class="m">3</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_111_0.png" src="../_images/1-Data_R_111_0.png" />
</div>
</div>
</div>
<div class="section" id="setting-transparency">
<h4>Setting transparency<a class="headerlink" href="#setting-transparency" title="Permalink to this headline">¶</a></h4>
<p>Because there are so many points, we can make them semi-transparent using <code class="docutils literal notranslate"><span class="pre">alpha</span></code> so that the overlaps can be seen:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="nf">I</span><span class="p">(</span><span class="n">.</span><span class="m">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_113_0.png" src="../_images/1-Data_R_113_0.png" />
</div>
</div>
<p>Here, try using <code class="docutils literal notranslate"><span class="pre">alpha</span> <span class="pre">=</span> <span class="pre">.5</span></code> instead of <code class="docutils literal notranslate"><span class="pre">alpha</span> <span class="pre">=</span> <span class="pre">I(.5)</span></code> and see what happens.</p>
</div>
<div class="section" id="adding-smoothers-and-regression-lines">
<h4>Adding smoothers and regression lines<a class="headerlink" href="#adding-smoothers-and-regression-lines" title="Permalink to this headline">¶</a></h4>
<p>Now add a smoother to the points:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;point&quot;</span><span class="p">,</span> <span class="s">&quot;smooth&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;
</pre></div>
</div>
<img alt="../_images/1-Data_R_115_1.png" src="../_images/1-Data_R_115_1.png" />
</div>
</div>
<p>If we want to have a linear regression, we need to specify the method as
being <code class="docutils literal notranslate"><span class="pre">lm</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;point&quot;</span><span class="p">,</span> <span class="s">&quot;smooth&quot;</span><span class="p">))</span> <span class="o">+</span> <span class="nf">geom_smooth</span><span class="p">(</span><span class="n">method</span> <span class="o">=</span> <span class="s">&quot;lm&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;

`geom_smooth()` using formula &#39;y ~ x&#39;
</pre></div>
</div>
<img alt="../_images/1-Data_R_117_1.png" src="../_images/1-Data_R_117_1.png" />
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">lm</span></code> stands for <code class="docutils literal notranslate"><span class="pre">l</span></code>inear <code class="docutils literal notranslate"><span class="pre">m</span></code>odels. Linear regression is a type of linear model. More on this in the <a class="reference internal" href="../Stats-Intro.html"><span class="doc std std-doc">BASIC DATA ANALYSES AND STATISTICS</span></a> section of these notes.</p>
<p>We can also add a “smoother” for each type of interaction:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;point&quot;</span><span class="p">,</span> <span class="s">&quot;smooth&quot;</span><span class="p">),</span> 
      <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_smooth</span><span class="p">(</span><span class="n">method</span> <span class="o">=</span> <span class="s">&quot;lm&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;

`geom_smooth()` using formula &#39;y ~ x&#39;
</pre></div>
</div>
<img alt="../_images/1-Data_R_119_1.png" src="../_images/1-Data_R_119_1.png" />
</div>
</div>
<p>To extend the lines to the full range, use <code class="docutils literal notranslate"><span class="pre">fullrange</span> <span class="pre">=</span> <span class="pre">TRUE</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;point&quot;</span><span class="p">,</span> <span class="s">&quot;smooth&quot;</span><span class="p">),</span>
      <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_smooth</span><span class="p">(</span><span class="n">method</span> <span class="o">=</span> <span class="s">&quot;lm&quot;</span><span class="p">,</span><span class="n">fullrange</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;

`geom_smooth()` using formula &#39;y ~ x&#39;
</pre></div>
</div>
<img alt="../_images/1-Data_R_121_1.png" src="../_images/1-Data_R_121_1.png" />
</div>
</div>
<p>Now let’s see how the ratio between prey and predator mass changes according to the type of interaction:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_123_0.png" src="../_images/1-Data_R_123_0.png" />
</div>
</div>
<p>Because there are so many points, we can “jitter” them to get a better
idea of the spread:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;jitter&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_125_0.png" src="../_images/1-Data_R_125_0.png" />
</div>
</div>
</div>
<div class="section" id="id2">
<h4>Boxplots<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h4>
<p>Or we can draw a boxplot of the data (note the <code class="docutils literal notranslate"><span class="pre">geom</span></code> argument, which stands for <code class="docutils literal notranslate"><span class="pre">geom</span></code>etry):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;boxplot&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_127_0.png" src="../_images/1-Data_R_127_0.png" />
</div>
</div>
</div>
<div class="section" id="histograms-and-density-plots">
<h4>Histograms and density plots<a class="headerlink" href="#histograms-and-density-plots" title="Permalink to this headline">¶</a></h4>
<p>Now let’s draw an histogram of predator-prey mass ratios:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span>  <span class="s">&quot;histogram&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
</pre></div>
</div>
<img alt="../_images/1-Data_R_129_1.png" src="../_images/1-Data_R_129_1.png" />
</div>
</div>
<p>Color the histogram according to the interaction type:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span>  <span class="s">&quot;histogram&quot;</span><span class="p">,</span> 
      <span class="n">fill</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
</pre></div>
</div>
<img alt="../_images/1-Data_R_131_1.png" src="../_images/1-Data_R_131_1.png" />
</div>
</div>
<p>You can also define your own  bin widths (in units of the x axis):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span>  <span class="s">&quot;histogram&quot;</span><span class="p">,</span> 
      <span class="n">fill</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="n">binwidth</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_133_0.png" src="../_images/1-Data_R_133_0.png" />
</div>
</div>
<p>To make it easier to read, we can plot the smoothed density of the data:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span>  <span class="s">&quot;density&quot;</span><span class="p">,</span> 
      <span class="n">fill</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_135_0.png" src="../_images/1-Data_R_135_0.png" />
</div>
</div>
<p>And you can make the densities transparent so that the overlaps are visible:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span>  <span class="s">&quot;density&quot;</span><span class="p">,</span> 
      <span class="n">fill</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">,</span> 
      <span class="n">alpha</span> <span class="o">=</span> <span class="nf">I</span><span class="p">(</span><span class="m">0.5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_137_0.png" src="../_images/1-Data_R_137_0.png" />
</div>
</div>
<p>or using <code class="docutils literal notranslate"><span class="pre">colour</span></code> instead of <code class="docutils literal notranslate"><span class="pre">fill</span></code> draws only the edge of the curve:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span>  <span class="s">&quot;density&quot;</span><span class="p">,</span> 
      <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_139_0.png" src="../_images/1-Data_R_139_0.png" />
</div>
</div>
<p>Similarly, <code class="docutils literal notranslate"><span class="pre">geom</span> <span class="pre">=</span> <span class="pre">&quot;bar&quot;</span></code> produces a barplot, <code class="docutils literal notranslate"><span class="pre">geom</span> <span class="pre">=</span> <span class="pre">&quot;line&quot;</span></code> a series of points joined by a line, etc.</p>
</div>
<div class="section" id="multi-faceted-plots">
<h4>Multi-faceted plots<a class="headerlink" href="#multi-faceted-plots" title="Permalink to this headline">¶</a></h4>
<p>An alternative way of displaying data belonging to different classes is using “faceting”:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">facets</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span> <span class="o">~</span><span class="n">.</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span>  <span class="s">&quot;density&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_141_0.png" src="../_images/1-Data_R_141_0.png" />
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">~.</span></code> (the space is not important) notation tells ggplot whether to do the faceting by row or by column. So if you want a by-column configuration, switch <code class="docutils literal notranslate"><span class="pre">~</span></code> and <code class="docutils literal notranslate"><span class="pre">.</span></code>, and also swap the position of the <code class="docutils literal notranslate"><span class="pre">.~</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">facets</span> <span class="o">=</span>  <span class="n">.</span><span class="o">~</span> <span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span>  <span class="s">&quot;density&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_143_0.png" src="../_images/1-Data_R_143_0.png" />
</div>
</div>
</div>
<div class="section" id="logarithmic-axes">
<h4>Logarithmic axes<a class="headerlink" href="#logarithmic-axes" title="Permalink to this headline">¶</a></h4>
<p>A better way to plot data in the log scale is to also set the axes to be logarithmic:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">,</span> <span class="n">Predator.mass</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="s">&quot;xy&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_145_0.png" src="../_images/1-Data_R_145_0.png" />
</div>
</div>
</div>
<div class="section" id="plot-annotations">
<h4>Plot annotations<a class="headerlink" href="#plot-annotations" title="Permalink to this headline">¶</a></h4>
<p>Let’s add a title and labels:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">,</span> <span class="n">Predator.mass</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="s">&quot;xy&quot;</span><span class="p">,</span>
    <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Relation between predator and prey mass&quot;</span><span class="p">,</span> 
    <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;log(Prey mass) (g)&quot;</span><span class="p">,</span> 
    <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;log(Predator mass) (g)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_147_0.png" src="../_images/1-Data_R_147_0.png" />
</div>
</div>
<p>Adding <code class="docutils literal notranslate"><span class="pre">+</span> <span class="pre">theme_bw()</span></code> makes it suitable for black and white printing.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">,</span> <span class="n">Predator.mass</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="s">&quot;xy&quot;</span><span class="p">,</span>
    <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Relation between predator and prey mass&quot;</span><span class="p">,</span> 
    <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;Prey mass (g)&quot;</span><span class="p">,</span> 
    <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Predator mass (g)&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme_bw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_149_0.png" src="../_images/1-Data_R_149_0.png" />
</div>
</div>
</div>
<div class="section" id="saving-your-plots">
<h4>Saving your plots<a class="headerlink" href="#saving-your-plots" title="Permalink to this headline">¶</a></h4>
<p>Finally, let’s save a pdf file of the figure (same approach as we <a class="reference external" href="#Saving-your-graphics">used before</a>):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">pdf</span><span class="p">(</span><span class="s">&quot;../results/MyFirst-ggplot2-Figure.pdf&quot;</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">qplot</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">,</span> <span class="n">Predator.mass</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span><span class="n">log</span><span class="o">=</span><span class="s">&quot;xy&quot;</span><span class="p">,</span>
    <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;Relation between predator and prey mass&quot;</span><span class="p">,</span> 
    <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;log(Prey mass) (g)&quot;</span><span class="p">,</span> 
    <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;log(Predator mass) (g)&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme_bw</span><span class="p">())</span>
<span class="nf">dev.off</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><strong>png:</strong> 2</div></div>
</div>
<p>Using <code class="docutils literal notranslate"><span class="pre">print</span></code> ensures that the whole command is kept together and that you can use the command in a script.</p>
</div>
</div>
<div class="section" id="some-more-important-ggplot-options">
<h3>Some more important ggplot options<a class="headerlink" href="#some-more-important-ggplot-options" title="Permalink to this headline">¶</a></h3>
<p>Other important options to keep in mind:</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Option</p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">xlim</span></code></p></td>
<td><p>limits for x axis: <code class="docutils literal notranslate"><span class="pre">xlim</span> <span class="pre">=</span> <span class="pre">c(0,12)</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">ylim</span></code></p></td>
<td><p>limits for y axis</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">log</span></code></p></td>
<td><p>log transform variable <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">&quot;x&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">&quot;y&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">&quot;xy&quot;</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">main</span></code></p></td>
<td><p>title of the plot <code class="docutils literal notranslate"><span class="pre">main</span> <span class="pre">=</span> <span class="pre">&quot;My</span> <span class="pre">Graph&quot;</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">xlab</span></code></p></td>
<td><p>x-axis label</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">ylab</span></code></p></td>
<td><p>y-axis label</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">asp</span></code></p></td>
<td><p>aspect ratio <code class="docutils literal notranslate"><span class="pre">asp</span> <span class="pre">=</span> <span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">asp</span> <span class="pre">=</span> <span class="pre">0.5</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">margins</span></code></p></td>
<td><p>whether or not margins will be displayed</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="the-geom-argument">
<h3>The <code class="docutils literal notranslate"><span class="pre">geom</span></code> argument<a class="headerlink" href="#the-geom-argument" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">geom</span></code> Specifies the geometric objects that define the graph type. The geom option is expressed as a R character vector with one or more entries. geom values include “point”, “smooth”, “boxplot”, “line”, “histogram”, “density”, “bar”, and “jitter”.</p>
<p>Try the following:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># load the data</span>
<span class="n">MyDF</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;../data/EcolArchives-E089-51-D1.csv&quot;</span><span class="p">))</span>

<span class="c1"># barplot</span>
<span class="nf">qplot</span><span class="p">(</span><span class="n">Predator.lifestage</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;bar&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_153_0.png" src="../_images/1-Data_R_153_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># boxplot</span>
<span class="nf">qplot</span><span class="p">(</span><span class="n">Predator.lifestage</span><span class="p">,</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;boxplot&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_154_0.png" src="../_images/1-Data_R_154_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># violin plot</span>
<span class="nf">qplot</span><span class="p">(</span><span class="n">Predator.lifestage</span><span class="p">,</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;violin&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_155_0.png" src="../_images/1-Data_R_155_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># density</span>
<span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;density&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_156_0.png" src="../_images/1-Data_R_156_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># histogram</span>
<span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;histogram&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
</pre></div>
</div>
<img alt="../_images/1-Data_R_157_1.png" src="../_images/1-Data_R_157_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># scatterplot</span>
<span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;point&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_158_0.png" src="../_images/1-Data_R_158_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># smooth</span>
<span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;smooth&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;
</pre></div>
</div>
<img alt="../_images/1-Data_R_159_1.png" src="../_images/1-Data_R_159_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">qplot</span><span class="p">(</span><span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">MyDF</span><span class="p">,</span> <span class="n">geom</span> <span class="o">=</span> <span class="s">&quot;smooth&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;lm&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`geom_smooth()` using formula &#39;y ~ x&#39;
</pre></div>
</div>
<img alt="../_images/1-Data_R_160_1.png" src="../_images/1-Data_R_160_1.png" />
</div>
</div>
</div>
<div class="section" id="advanced-plotting-ggplot">
<h3>Advanced plotting: <code class="docutils literal notranslate"><span class="pre">ggplot</span></code><a class="headerlink" href="#advanced-plotting-ggplot" title="Permalink to this headline">¶</a></h3>
<p>The command <code class="docutils literal notranslate"><span class="pre">qplot</span></code> allows you to use only a single dataset and a single set of “aesthetics” (x, y, etc.). To make full use of <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code>, we need to use the command <code class="docutils literal notranslate"><span class="pre">ggplot</span></code>, which allows you to use “layering”. Layering is the mechanism by which additional data elements are added to a plot. Each layer can come from a different dataset and have a different aesthetic mapping, allowing us to create plots that could not be generated using <code class="docutils literal notranslate"><span class="pre">qplot()</span></code>, which permits only a single dataset and a single set of aesthetic mappings.</p>
<p>For a <code class="docutils literal notranslate"><span class="pre">ggplot</span></code> plotting command, we need at least:</p>
<ul class="simple">
<li><p>The data to be plotted, in a data frame;</p></li>
<li><p>Aesthetics mappings, specifying which variables we want to plot, and how;</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">geom</span></code>, defining the geometry for representing the data;</p></li>
<li><p>(Optionally) some <code class="docutils literal notranslate"><span class="pre">stat</span></code> that transforms the data or performs statistics using the data.</p></li>
</ul>
<p>To start a graph, we must specify the data and the aesthetics:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span>
                <span class="n">y</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span>
                <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>Here we have created a graphics object <code class="docutils literal notranslate"><span class="pre">p</span></code> to which we can add layers and other plot elements.</p>
<p>Now try to plot the graph:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_164_0.png" src="../_images/1-Data_R_164_0.png" />
</div>
</div>
<p>Plot is blank because we are yet to specify a geometry — only then can we see the graph:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">geom_point</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_166_0.png" src="../_images/1-Data_R_166_0.png" />
</div>
</div>
<p>We can use the “+” sign to concatenate different commands:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span> <span class="p">))</span>
<span class="n">q</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> 
<span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="nf">I</span><span class="p">(</span><span class="m">2</span><span class="p">),</span> <span class="n">shape</span><span class="o">=</span><span class="nf">I</span><span class="p">(</span><span class="m">10</span><span class="p">))</span> <span class="o">+</span>
<span class="nf">theme_bw</span><span class="p">()</span> <span class="o">+</span> <span class="c1"># make the background white</span>
<span class="nf">theme</span><span class="p">(</span><span class="n">aspect.ratio</span><span class="o">=</span><span class="m">1</span><span class="p">)</span> <span class="c1"># make the plot square</span>
<span class="n">q</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_168_0.png" src="../_images/1-Data_R_168_0.png" />
</div>
</div>
<p>Let’s remove the legend:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">q</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">aspect.ratio</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_170_0.png" src="../_images/1-Data_R_170_0.png" />
</div>
</div>
<p>Now, same as before, but using ggplot instead of qplot:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span> <span class="p">))</span> <span class="o">+</span> <span class="nf">geom_density</span><span class="p">()</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_172_0.png" src="../_images/1-Data_R_172_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span><span class="p">))</span> <span class="o">+</span> <span class="nf">geom_density</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="m">0.5</span><span class="p">)</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_173_0.png" src="../_images/1-Data_R_173_0.png" />
</div>
</div>
<p>You can also create a multifaceted plot:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span> <span class="p">))</span> <span class="o">+</span>  <span class="nf">geom_density</span><span class="p">()</span> <span class="o">+</span> <span class="nf">facet_wrap</span><span class="p">(</span> <span class="n">.</span><span class="o">~</span> <span class="n">Type.of.feeding.interaction</span><span class="p">)</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_176_0.png" src="../_images/1-Data_R_176_0.png" />
</div>
</div>
<p>The axis limits in the plots are fixed irrespective of the data ranges, making it harder to visualize the shapes of the distributions. You can free up the axes to allow data-specific axis limits by using the <code class="docutils literal notranslate"><span class="pre">scales</span> <span class="pre">=</span> <span class="pre">&quot;free&quot;</span></code> argument:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span> <span class="p">))</span> <span class="o">+</span>  <span class="nf">geom_density</span><span class="p">()</span> <span class="o">+</span> <span class="nf">facet_wrap</span><span class="p">(</span> <span class="n">.</span><span class="o">~</span> <span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="n">scales</span> <span class="o">=</span> <span class="s">&quot;free&quot;</span><span class="p">)</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_178_0.png" src="../_images/1-Data_R_178_0.png" />
</div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>In <code class="docutils literal notranslate"><span class="pre">facte_wrap</span></code>, You can also free up just the x or y scales; look up the documentation for this function.</p>
</div>
<p>Now let’s plot the size-ratio distributions by location (and again, with scales free):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="o">/</span><span class="n">Predator.mass</span><span class="p">)))</span> <span class="o">+</span>  <span class="nf">geom_density</span><span class="p">()</span> <span class="o">+</span> <span class="nf">facet_wrap</span><span class="p">(</span> <span class="n">.</span><span class="o">~</span> <span class="n">Location</span><span class="p">,</span> <span class="n">scales</span> <span class="o">=</span> <span class="s">&quot;free&quot;</span><span class="p">)</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_182_0.png" src="../_images/1-Data_R_182_0.png" />
</div>
</div>
<p>A different example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">)))</span> <span class="o">+</span>  <span class="nf">geom_point</span><span class="p">()</span> <span class="o">+</span> <span class="nf">facet_wrap</span><span class="p">(</span> <span class="n">.</span><span class="o">~</span> <span class="n">Location</span><span class="p">,</span> <span class="n">scales</span> <span class="o">=</span> <span class="s">&quot;free&quot;</span><span class="p">)</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_184_0.png" src="../_images/1-Data_R_184_0.png" />
</div>
</div>
<p>You can also combine categories like this (this will be BIG plot!):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">)))</span> <span class="o">+</span>  <span class="nf">geom_point</span><span class="p">()</span> <span class="o">+</span> <span class="nf">facet_wrap</span><span class="p">(</span> <span class="n">.</span><span class="o">~</span> <span class="n">Location</span> <span class="o">+</span> <span class="n">Type.of.feeding.interaction</span><span class="p">,</span> <span class="n">scales</span> <span class="o">=</span> <span class="s">&quot;free&quot;</span><span class="p">)</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_187_0.png" src="../_images/1-Data_R_187_0.png" />
</div>
</div>
<p>And you can also change the order of the combination:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">)))</span> <span class="o">+</span>  <span class="nf">geom_point</span><span class="p">()</span> <span class="o">+</span> <span class="nf">facet_wrap</span><span class="p">(</span> <span class="n">.</span><span class="o">~</span> <span class="n">Type.of.feeding.interaction</span> <span class="o">+</span> <span class="n">Location</span><span class="p">,</span> <span class="n">scales</span> <span class="o">=</span> <span class="s">&quot;free&quot;</span><span class="p">)</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_189_0.png" src="../_images/1-Data_R_189_0.png" />
</div>
</div>
<p>For more fine-tuned faceting, look up the <code class="docutils literal notranslate"><span class="pre">facet_grid()</span></code> and <code class="docutils literal notranslate"><span class="pre">facet_wrap()</span></code> functions within <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code>. Look up this variant of the <a class="reference external" href="http://www.cookbook-r.com/Graphs">R Cookbook</a> for more examples.</p>
</div>
<div class="section" id="some-useful-ggplot-examples">
<h3>Some useful ggplot examples<a class="headerlink" href="#some-useful-ggplot-examples" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="plotting-a-matrix">
<h3>Plotting a matrix<a class="headerlink" href="#plotting-a-matrix" title="Permalink to this headline">¶</a></h3>
<p>Let’s plot the values of a matrix. This is basically the same as rendering a 2D image. We will visualize random values taken from a unform distribution <span class="math notranslate nohighlight">\(\mathcal U [0,1]\)</span>. Because we want to plot a matrix, and ggplot2 accepts only dataframes, we use the package reshape2, which can “melt” a matrix into a dataframe:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">require</span><span class="p">(</span><span class="n">reshape2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">GenerateMatrix</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">N</span><span class="p">){</span>
    <span class="n">M</span> <span class="o">&lt;-</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">runif</span><span class="p">(</span><span class="n">N</span> <span class="o">*</span> <span class="n">N</span><span class="p">),</span> <span class="n">N</span><span class="p">,</span> <span class="n">N</span><span class="p">)</span>
    <span class="nf">return</span><span class="p">(</span><span class="n">M</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">M</span> <span class="o">&lt;-</span> <span class="nf">GenerateMatrix</span><span class="p">(</span><span class="m">10</span><span class="p">)</span>
<span class="n">Melt</span> <span class="o">&lt;-</span> <span class="nf">melt</span><span class="p">(</span><span class="n">M</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">Melt</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">Var1</span><span class="p">,</span> <span class="n">Var2</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">value</span><span class="p">))</span> <span class="o">+</span> <span class="nf">geom_tile</span><span class="p">()</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">aspect.ratio</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_196_0.png" src="../_images/1-Data_R_196_0.png" />
</div>
</div>
<p>Add a black line dividing cells:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">geom_tile</span><span class="p">(</span><span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">aspect.ratio</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_198_0.png" src="../_images/1-Data_R_198_0.png" />
</div>
</div>
<p>Remove the legend:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">aspect.ratio</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_200_0.png" src="../_images/1-Data_R_200_0.png" />
</div>
</div>
<p>Remove all the rest:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">,</span> 
     <span class="n">panel.background</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
     <span class="n">axis.ticks</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span> 
     <span class="n">panel.grid.major</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
     <span class="n">panel.grid.minor</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
     <span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
     <span class="n">axis.title.x</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
     <span class="n">axis.text.y</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
     <span class="n">axis.title.y</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_202_0.png" src="../_images/1-Data_R_202_0.png" />
</div>
</div>
<p>Explore some colors:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">scale_fill_continuous</span><span class="p">(</span><span class="n">low</span> <span class="o">=</span> <span class="s">&quot;yellow&quot;</span><span class="p">,</span> <span class="n">high</span> <span class="o">=</span> <span class="s">&quot;darkgreen&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_204_0.png" src="../_images/1-Data_R_204_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">scale_fill_gradient2</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_205_0.png" src="../_images/1-Data_R_205_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">scale_fill_gradientn</span><span class="p">(</span><span class="n">colours</span> <span class="o">=</span> <span class="nf">grey.colors</span><span class="p">(</span><span class="m">10</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_206_0.png" src="../_images/1-Data_R_206_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">scale_fill_gradientn</span><span class="p">(</span><span class="n">colours</span> <span class="o">=</span> <span class="nf">rainbow</span><span class="p">(</span><span class="m">10</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_207_0.png" src="../_images/1-Data_R_207_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">+</span> <span class="nf">scale_fill_gradientn</span><span class="p">(</span><span class="n">colours</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="p">,</span> <span class="s">&quot;white&quot;</span><span class="p">,</span> <span class="s">&quot;blue&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_208_0.png" src="../_images/1-Data_R_208_0.png" />
</div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Look up the <code class="docutils literal notranslate"><span class="pre">plotly</span></code> package, which allows you to create interactive graphs from ggplot2 plots</p>
</div>
</div>
<div class="section" id="plotting-two-dataframes-together">
<h3>Plotting two dataframes together<a class="headerlink" href="#plotting-two-dataframes-together" title="Permalink to this headline">¶</a></h3>
<p>According to Girko’s circular law, the eigenvalues of a matrix <span class="math notranslate nohighlight">\(M\)</span> of size <span class="math notranslate nohighlight">\(N \times N\)</span> are approximately contained in a circle in the complex plane with radius <span class="math notranslate nohighlight">\(\sqrt{N}\)</span>. Let’s draw the results of a simulation displaying this result.</p>
<p>First, we need to build a function object that will calculate the ellipse (the perdicted bounds of the eigenvalues):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">build_ellipse</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">hradius</span><span class="p">,</span> <span class="n">vradius</span><span class="p">){</span> <span class="c1"># function that returns an ellipse</span>
  <span class="n">npoints</span> <span class="o">=</span> <span class="m">250</span>
  <span class="n">a</span> <span class="o">&lt;-</span> <span class="nf">seq</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">2</span> <span class="o">*</span> <span class="kc">pi</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="n">npoints</span> <span class="o">+</span> <span class="m">1</span><span class="p">)</span>
  <span class="n">x</span> <span class="o">&lt;-</span> <span class="n">hradius</span> <span class="o">*</span> <span class="nf">cos</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
  <span class="n">y</span> <span class="o">&lt;-</span> <span class="n">vradius</span> <span class="o">*</span> <span class="nf">sin</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>  
  <span class="nf">return</span><span class="p">(</span><span class="nf">data.frame</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">y</span><span class="p">))</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">&lt;-</span> <span class="m">250</span> <span class="c1"># Assign size of the matrix</span>

<span class="n">M</span> <span class="o">&lt;-</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">rnorm</span><span class="p">(</span><span class="n">N</span> <span class="o">*</span> <span class="n">N</span><span class="p">),</span> <span class="n">N</span><span class="p">,</span> <span class="n">N</span><span class="p">)</span> <span class="c1"># Build the matrix</span>

<span class="n">eigvals</span> <span class="o">&lt;-</span> <span class="nf">eigen</span><span class="p">(</span><span class="n">M</span><span class="p">)</span><span class="o">$</span><span class="n">values</span> <span class="c1"># Find the eigenvalues</span>

<span class="n">eigDF</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="s">&quot;Real&quot;</span> <span class="o">=</span> <span class="nf">Re</span><span class="p">(</span><span class="n">eigvals</span><span class="p">),</span> <span class="s">&quot;Imaginary&quot;</span> <span class="o">=</span> <span class="nf">Im</span><span class="p">(</span><span class="n">eigvals</span><span class="p">))</span> <span class="c1"># Build a dataframe</span>

<span class="n">my_radius</span> <span class="o">&lt;-</span> <span class="nf">sqrt</span><span class="p">(</span><span class="n">N</span><span class="p">)</span> <span class="c1"># The radius of the circle is sqrt(N)</span>

<span class="n">ellDF</span> <span class="o">&lt;-</span> <span class="nf">build_ellipse</span><span class="p">(</span><span class="n">my_radius</span><span class="p">,</span> <span class="n">my_radius</span><span class="p">)</span> <span class="c1"># Dataframe to plot the ellipse</span>

<span class="nf">names</span><span class="p">(</span><span class="n">ellDF</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Real&quot;</span><span class="p">,</span> <span class="s">&quot;Imaginary&quot;</span><span class="p">)</span> <span class="c1"># rename the columns</span>
</pre></div>
</div>
</div>
</div>
<p>Now the plotting:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot the eigenvalues</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">eigDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">Real</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">Imaginary</span><span class="p">))</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">shape</span> <span class="o">=</span> <span class="nf">I</span><span class="p">(</span><span class="m">3</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span>

<span class="c1"># now add the vertical and horizontal line</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_hline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="m">0</span><span class="p">))</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_vline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="m">0</span><span class="p">))</span>

<span class="c1"># finally, add the ellipse</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_polygon</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">ellDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">Real</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">Imaginary</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="m">1</span><span class="o">/</span><span class="m">20</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s">&quot;red&quot;</span><span class="p">))</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_214_0.png" src="../_images/1-Data_R_214_0.png" />
</div>
</div>
<div class="section" id="id3">
<h4>Exercise<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h4>
<p>Create a self-sufficient R Script called <code class="docutils literal notranslate"><span class="pre">Girko.R</span></code> that combines all the above commands for plotting the Girko’s law simulation, including saving the result figure as a file called <code class="docutils literal notranslate"><span class="pre">Girko.pdf</span></code> in the results directory.</p>
</div>
</div>
<div class="section" id="annotating-plots">
<h3>Annotating plots<a class="headerlink" href="#annotating-plots" title="Permalink to this headline">¶</a></h3>
<p>In this example, we will use the ggplot <code class="docutils literal notranslate"><span class="pre">geom</span></code> <code class="docutils literal notranslate"><span class="pre">text</span></code> to annotate a plot.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">a</span> <span class="o">&lt;-</span> <span class="nf">read.table</span><span class="p">(</span><span class="s">&quot;../data/Results.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Check what the data look like:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A data.frame: 6 × 5</caption>
<thead>
	<tr><th></th><th scope=col>x</th><th scope=col>y1</th><th scope=col>y2</th><th scope=col>y3</th><th scope=col>Label</th></tr>
	<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>3.515424</td><td>4320</td><td>4320</td><td>0</td><td>NA</td></tr>
	<tr><th scope=row>2</th><td>3.533984</td><td>2160</td><td>2160</td><td>0</td><td>NA</td></tr>
	<tr><th scope=row>3</th><td>3.557647</td><td>4320</td><td>4320</td><td>0</td><td>NA</td></tr>
	<tr><th scope=row>4</th><td>3.569953</td><td>4320</td><td>4320</td><td>0</td><td>NA</td></tr>
	<tr><th scope=row>5</th><td>3.578984</td><td>8640</td><td>8640</td><td>0</td><td>NA</td></tr>
	<tr><th scope=row>6</th><td>3.585665</td><td>2160</td><td>2160</td><td>0</td><td>NA</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">a</span><span class="o">$</span><span class="n">ymin</span> <span class="o">&lt;-</span> <span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="nf">dim</span><span class="p">(</span><span class="n">a</span><span class="p">)[</span><span class="m">1</span><span class="p">])</span> <span class="c1"># append a column of zeros</span>

<span class="c1"># Print the first linerange</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_linerange</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">a</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span>
                          <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span>
                          <span class="n">ymin</span> <span class="o">=</span> <span class="n">ymin</span><span class="p">,</span>
                          <span class="n">ymax</span> <span class="o">=</span> <span class="n">y1</span><span class="p">,</span>
                          <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="m">0.5</span><span class="p">)</span>
                          <span class="p">),</span>
                        <span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;#E69F00&quot;</span><span class="p">,</span>
                        <span class="n">alpha</span> <span class="o">=</span> <span class="m">1</span><span class="o">/</span><span class="m">2</span><span class="p">,</span> <span class="n">show.legend</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="c1"># Print the second linerange</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_linerange</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">a</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span>
                          <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span>
                          <span class="n">ymin</span> <span class="o">=</span> <span class="n">ymin</span><span class="p">,</span>
                          <span class="n">ymax</span> <span class="o">=</span> <span class="n">y2</span><span class="p">,</span>
                          <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="m">0.5</span><span class="p">)</span>
                          <span class="p">),</span>
                        <span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;#56B4E9&quot;</span><span class="p">,</span>
                        <span class="n">alpha</span> <span class="o">=</span> <span class="m">1</span><span class="o">/</span><span class="m">2</span><span class="p">,</span> <span class="n">show.legend</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="c1"># Print the third linerange:</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_linerange</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">a</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span>
                          <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span>
                          <span class="n">ymin</span> <span class="o">=</span> <span class="n">ymin</span><span class="p">,</span>
                          <span class="n">ymax</span> <span class="o">=</span> <span class="n">y3</span><span class="p">,</span>
                          <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="m">0.5</span><span class="p">)</span>
                          <span class="p">),</span>
                        <span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;#D55E00&quot;</span><span class="p">,</span>
                        <span class="n">alpha</span> <span class="o">=</span> <span class="m">1</span><span class="o">/</span><span class="m">2</span><span class="p">,</span> <span class="n">show.legend</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="c1"># Annotate the plot with labels:</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_text</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">a</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="m">-500</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">Label</span><span class="p">))</span>

<span class="c1"># now set the axis labels, remove the legend, and prepare for bw printing</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">scale_x_continuous</span><span class="p">(</span><span class="s">&quot;My x axis&quot;</span><span class="p">,</span>
                            <span class="n">breaks</span> <span class="o">=</span> <span class="nf">seq</span><span class="p">(</span><span class="m">3</span><span class="p">,</span> <span class="m">5</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">))</span> <span class="o">+</span> 
                            <span class="nf">scale_y_continuous</span><span class="p">(</span><span class="s">&quot;My y axis&quot;</span><span class="p">)</span> <span class="o">+</span> 
                            <span class="nf">theme_bw</span><span class="p">()</span> <span class="o">+</span> 
                            <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span> 
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message:
“Removed 91 rows containing missing values (geom_text).”
</pre></div>
</div>
<img alt="../_images/1-Data_R_220_1.png" src="../_images/1-Data_R_220_1.png" />
</div>
</div>
</div>
</div>
<div class="section" id="id4">
<h2>Practicals<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h2>
<p>Create a self-sufficient R Script called <code class="docutils literal notranslate"><span class="pre">MyBars.pdf</span></code> that combines all the above commands for annotating a plot, including saving the result figure as a file called <code class="docutils literal notranslate"><span class="pre">MyBars.pdf</span></code> in the results directory.</p>
</div>
<div class="section" id="mathematical-display">
<h2>Mathematical display<a class="headerlink" href="#mathematical-display" title="Permalink to this headline">¶</a></h2>
<p>Let’s try mathematical annotation on a axis, and in the plot area.</p>
<p>First create some linear regression “data”:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">&lt;-</span> <span class="nf">seq</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">100</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="m">0.1</span><span class="p">)</span>
<span class="n">y</span> <span class="o">&lt;-</span> <span class="m">-4</span><span class="n">.</span> <span class="o">+</span> <span class="m">0.25</span> <span class="o">*</span> <span class="n">x</span> <span class="o">+</span>
  <span class="nf">rnorm</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">mean</span> <span class="o">=</span> <span class="m">0</span><span class="n">.</span><span class="p">,</span> <span class="n">sd</span> <span class="o">=</span> <span class="m">2.5</span><span class="p">)</span>

<span class="c1"># and put them in a dataframe</span>
<span class="n">my_data</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">y</span><span class="p">)</span>

<span class="c1"># perform a linear regression</span>
<span class="n">my_lm</span> <span class="o">&lt;-</span> <span class="nf">summary</span><span class="p">(</span><span class="nf">lm</span><span class="p">(</span><span class="n">y</span> <span class="o">~</span> <span class="n">x</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">my_data</span><span class="p">))</span>

<span class="c1"># plot the data</span>
<span class="n">p</span> <span class="o">&lt;-</span>  <span class="nf">ggplot</span><span class="p">(</span><span class="n">my_data</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">y</span><span class="p">,</span>
                          <span class="n">colour</span> <span class="o">=</span> <span class="nf">abs</span><span class="p">(</span><span class="n">my_lm</span><span class="o">$</span><span class="n">residual</span><span class="p">))</span>
             <span class="p">)</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">scale_colour_gradient</span><span class="p">(</span><span class="n">low</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">,</span> <span class="n">high</span> <span class="o">=</span> <span class="s">&quot;red&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_x_continuous</span><span class="p">(</span>
    <span class="nf">expression</span><span class="p">(</span><span class="n">alpha</span><span class="o">^</span><span class="m">2</span> <span class="o">*</span> <span class="kc">pi</span> <span class="o">/</span> <span class="n">beta</span> <span class="o">*</span> <span class="nf">sqrt</span><span class="p">(</span><span class="n">Theta</span><span class="p">)))</span>

<span class="c1"># add the regression line</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_abline</span><span class="p">(</span>
  <span class="n">intercept</span> <span class="o">=</span> <span class="n">my_lm</span><span class="o">$</span><span class="n">coefficients</span><span class="p">[</span><span class="m">1</span><span class="p">][</span><span class="m">1</span><span class="p">],</span>
  <span class="n">slope</span> <span class="o">=</span> <span class="n">my_lm</span><span class="o">$</span><span class="n">coefficients</span><span class="p">[</span><span class="m">2</span><span class="p">][</span><span class="m">1</span><span class="p">],</span>
  <span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;red&quot;</span><span class="p">)</span>
<span class="c1"># throw some math on the plot</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="m">60</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span>
                       <span class="n">label</span> <span class="o">=</span> <span class="s">&quot;sqrt(alpha) * 2* pi&quot;</span><span class="p">),</span> 
                       <span class="n">parse</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="m">6</span><span class="p">,</span> 
                       <span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;blue&quot;</span><span class="p">)</span>

<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_223_0.png" src="../_images/1-Data_R_223_0.png" />
</div>
</div>
</div>
<div class="section" id="ggthemes">
<h2>ggthemes<a class="headerlink" href="#ggthemes" title="Permalink to this headline">¶</a></h2>
<p>The package <code class="docutils literal notranslate"><span class="pre">ggthemes</span></code> provides you some additional <code class="docutils literal notranslate"><span class="pre">geom</span></code>s, <code class="docutils literal notranslate"> <span class="pre">scale</span></code>s, and <code class="docutils literal notranslate"><span class="pre">theme</span></code>s for <code class="docutils literal notranslate"><span class="pre">ggplot</span></code>. These include a theme based on Tufte’s <em>The Visual Display of Quantitative Information</em> (see <a class="reference external" href="#Readings-&amp;-Resources">suggested readings</a>).</p>
<p>Let’s try this package. First install it:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;ggthemes&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Then try:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">ggthemes</span><span class="p">)</span>

<span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">MyDF</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Predator.mass</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">Prey.mass</span><span class="p">),</span>
                <span class="n">colour</span> <span class="o">=</span> <span class="n">Type.of.feeding.interaction</span> <span class="p">))</span> <span class="o">+</span>
                <span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="nf">I</span><span class="p">(</span><span class="m">2</span><span class="p">),</span> <span class="n">shape</span><span class="o">=</span><span class="nf">I</span><span class="p">(</span><span class="m">10</span><span class="p">))</span> <span class="o">+</span> <span class="nf">theme_bw</span><span class="p">()</span>

<span class="n">p</span> <span class="o">+</span> <span class="nf">geom_rangeframe</span><span class="p">()</span> <span class="o">+</span> <span class="c1"># now fine tune the geom to Tufte&#39;s range frame</span>
        <span class="nf">theme_tufte</span><span class="p">()</span> <span class="c1"># and theme to Tufte&#39;s minimal ink theme    </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1-Data_R_225_0.png" src="../_images/1-Data_R_225_0.png" />
</div>
</div>
<p>Go to &lt;https://github.com/jrnold/ggthemes for more information and a list of <code class="docutils literal notranslate"><span class="pre">geom</span></code>s, <code class="docutils literal notranslate"><span class="pre">theme</span></code>s, and <code class="docutils literal notranslate"><span class="pre">scale</span></code>s.</p>
</div>
<div class="section" id="id5">
<h2>Practicals<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h2>
<div class="section" id="plotting-script">
<h3>Plotting script<a class="headerlink" href="#plotting-script" title="Permalink to this headline">¶</a></h3>
<p>Create a self-sufficient R Script called <code class="docutils literal notranslate"><span class="pre">plotLin.R</span></code> that combines all the above commands for annotating a plot, including saving the resulting figure as a file called <code class="docutils literal notranslate"><span class="pre">MyLinReg.pdf</span></code> in the results directory.</p>
</div>
<div class="section" id="visualizing-regression-analyses">
<h3>Visualizing Regression analyses<a class="headerlink" href="#visualizing-regression-analyses" title="Permalink to this headline">¶</a></h3>
<p>In this practical, you will write script that draws and saves a pdf file of the following figure, and writes the accompanying regression results to a formatted table in csv. Note that as the plot suggests, that the analysis must be subsetted by the <code class="docutils literal notranslate"><span class="pre">Predator.lifestage</span></code> field of the dataset.</p>
<hr class="docutils" />
<p><img alt="image" src="../_images/Figure1.png" />
<small><center>
Write a script that generates this figure
</center></small></p>
<p><em>Guidelines</em>:</p>
<ul class="simple">
<li><p>Write a <code class="docutils literal notranslate"><span class="pre">R</span></code> script file called <code class="docutils literal notranslate"><span class="pre">PP_Regress.R</span></code> and save it in the <code class="docutils literal notranslate"><span class="pre">Code</span></code> directory. Sourcing or running this
script should result in one pdf file containing the following figure being saved in the <code class="docutils literal notranslate"><span class="pre">results</span></code> directory: (Hint: Use the <code class="docutils literal notranslate"><span class="pre">print()</span></code> command to write to the pdf)</p></li>
<li><p>In addition, the script should calculate the regression results corresponding to the lines fitted in the figure and save it to a csv delimited table called (<code class="docutils literal notranslate"><span class="pre">PP_Regress_Results.csv</span></code>), in the <code class="docutils literal notranslate"><span class="pre">results</span></code> directory. (Hint: you will have to initialize a new dataframe in the script to first store the calculations and then <code class="docutils literal notranslate"><span class="pre">write.csv()</span></code> or <code class="docutils literal notranslate"><span class="pre">write.table()</span></code> it.)</p></li>
<li><p>All that you are being asked for here is results of an analysis of Linear regression on subsets of the data corresponding to available Feeding Type <span class="math notranslate nohighlight">\(\times\)</span> Predator life Stage combination — not a multivariate linear model with these two as separate covariates!</p></li>
<li><p>The regression results should include the following with appropriate headers (e.g., slope, intercept, etc, in each Feeding type <span class="math notranslate nohighlight">\(\times\)</span> life stage category): regression slope, regression intercept, R<span class="math notranslate nohighlight">\(^2\)</span>, F-statistic value, and p-value of the overall regression (Hint: Review the Stats week!).</p></li>
<li><p>The script should be self-sufficient and not need any external inputs — it should import the above predator-prey dataset from the appropriate directory, and save the graphic plots to the appropriate directory (Hint: use relative paths). Anybody should be able to <code class="docutils literal notranslate"><span class="pre">source</span></code> it without errors.</p></li>
<li><p>You can also use the <code class="docutils literal notranslate"><span class="pre">dplyr</span></code> function instead of looping (see R Chapter, and the <code class="docutils literal notranslate"><span class="pre">ggplot</span></code> command instead of <code class="docutils literal notranslate"><span class="pre">qplot</span></code>.</p></li>
</ul>
</div>
<div class="section" id="mapping">
<h3>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h3>
<p>You can do mapping in R using the <code class="docutils literal notranslate"><span class="pre">maps</span></code> package. In this practical, you will map the Global Population Dynamics Database (GPDD). This is a freely available database that was developed at Silwood. The <a class="reference external" href="http://livingplanetindex.org/home/index">Living Planet Index</a> is based upon these data.</p>
<ul class="simple">
<li><p>Use <code class="docutils literal notranslate"><span class="pre">load()</span></code> from <code class="docutils literal notranslate"><span class="pre">GPDDFiltered.RData</span></code> that is available on the TheMulQuaBio git repository — have a look at the data contents.</p></li>
<li><p>What you need is latitude and longitude information for a bunch of species for which population time series are available in the GPDD.</p></li>
<li><p>Now use <code class="docutils literal notranslate"><span class="pre">install.packages()</span></code> to install the package <code class="docutils literal notranslate"><span class="pre">maps</span></code>, as you did with <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code></p></li>
<li><p>Now create a script called <code class="docutils literal notranslate"><span class="pre">GPDD_Data.R</span></code> that:</p>
<ul>
<li><p>Loads the maps package</p></li>
<li><p>Loads the GPDD data</p></li>
<li><p>Creates a world map (use the map function, read its help, also google examples using <code class="docutils literal notranslate"><span class="pre">maps</span></code></p></li>
<li><p>Superimposes on the map all the locations from which we have data in the GPDD dataframe</p></li>
</ul>
</li>
<li><p>Looking at the map, what biases might you expect in any analysis based on the data represented? <em>include your answer as a comment at the end of your R script</em>.</p></li>
</ul>
</div>
<div class="section" id="groupwork-practical-regression-analysis">
<h3>Groupwork practical: Regression analysis<a class="headerlink" href="#groupwork-practical-regression-analysis" title="Permalink to this headline">¶</a></h3>
<p>Do the same as (1) above, but the analysis this time should be separate by the dataset’s <code class="docutils literal notranslate"><span class="pre">Location</span></code> field. Call it <code class="docutils literal notranslate"><span class="pre">PP_Regress_loc.R</span></code>. No need to generate plots for this (just the analysis results to a <code class="docutils literal notranslate"><span class="pre">.csv</span></code> file), as a combination of <code class="docutils literal notranslate"><span class="pre">Type.of.feeding.interaction</span></code>, <code class="docutils literal notranslate"><span class="pre">Predator.lifestage</span></code>, and <code class="docutils literal notranslate"> <span class="pre">Location</span></code> will be far too busy (faceting by three variables is too much to be meaningful)!</p>
</div>
</div>
<div class="section" id="readings-resources">
<h2>Readings &amp; Resources<a class="headerlink" href="#readings-resources" title="Permalink to this headline">¶</a></h2>
<p>Check out <em>DataDataData!</em>, <em>Visualization</em> and <em>R</em>, under <a class="reference external" href="https://github.com/mhasoba/TheMulQuaBio/tree/master/content/readings"><em>readings</em> on the TheMulQuaBio repository</a>.</p>
<div class="section" id="data-management-reformatting-and-cleaning">
<h3>Data management, reformatting and cleaning<a class="headerlink" href="#data-management-reformatting-and-cleaning" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Brian McGill’s <a class="reference external" href="https://dynamicecology.wordpress.com/2016/08/22/ten-commandments-for-good-data-management">Ten commandments for data management</a></p></li>
<li><p>This paper covers similar ground (available in the readings directory): <a class="reference external" href="https://esajournals.onlinelibrary.wiley.com/doi/full/10.1890/0012-9623-90.2.205">Borer et al (2009). Some Simple Guidelines for Effective Data Management. Bulletin of the Ecological Society of America, 90(2), 205-214.</a></p></li>
<li><p><a class="reference external" href="https://onlinelibrary.wiley.com/doi/book/10.1002/0471448354">Dasu, T. &amp; Johnson, T. Exploratory Data Mining and Data Cleaning. (John Wiley &amp; Sons, Inc., 2003). doi:10.1002/0471448354</a></p></li>
<li><p><a class="reference external" href="http://www.theanalysisfactor.com/wide-and-long-data/">Wide vs. long data</a></p></li>
<li><p><a class="reference external" href="http://vita.had.co.nz/papers/tidy-data.pdf">Hadley Wickham’s excellent article about “tidy” data</a>.</p></li>
<li><p><a class="reference external" href="http://vis.stanford.edu/papers/wrangler">Wrangler</a></p></li>
<li><p><a class="reference external" href="https://www2.eecs.berkeley.edu/Pubs/TechRpts/2000/CSD-00-1110.pdf">An interactive framework for data cleaning</a></p></li>
</ul>
</div>
<div class="section" id="visualization">
<h3>Visualization<a class="headerlink" href="#visualization" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003833">Ten Simple Rules for Better Figures</a></p></li>
<li><p>https://clauswilke.com/dataviz/</p></li>
<li><p><a class="reference external" href="https://onlinelibrary.wiley.com/doi/pdf/10.1002/adma.201102518?casa_token=48b0PGtaUiYAAAAA:DQ9bHugikOkHDkf4uLTYIq74negyWefZjAbexYcgYObHn0sks2Q8ulvHcTUWxteqwTb5qAa0IV71Mg">Rolandi et al. “A Brief Guide to Designing Effective Figures for the Scientific Paper”, doi:10.1002/adma.201102518</a></p></li>
<li><p>The classic <a class="reference external" href="https://www.edwardtufte.com/tufte/books_vdqi">Tufte</a>; Available in the Imperial College Central Library. I have also added extracts and a related book in pdf on the git repository. BTW, check out what Tufte thinks of <a class="reference external" href="https://www.edwardtufte.com/tufte/powerpoint">PowerPoint</a>.</p></li>
<li><p><a class="reference external" href="http://motioninsocial.com/tufte">Tufte in R</a></p></li>
</ul>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="0-intro.html" title="previous page">Welcome to the VectorBiTE 2021 Training Workshop!</a>
    <a class='right-next' id="next-link" href="2-Digitizer_explanation.html" title="next page">How to Digitize Data from Graphs in Published Papers</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By VectorBiTE Team<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>